// @govtechsg/sgds-react/Tabs/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var PropTypes = require('prop-types');
var uncontrollable = require('uncontrollable');
var BaseTabs = require('@restart/ui/Tabs');
var classNames = require('classnames');
var BaseNav = require('@restart/ui/Nav');
var createWithBsPrefix = require('../utils/createWithBsPrefix');
var Anchor = require('@restart/ui/Anchor');
var NavItem$1 = require('@restart/ui/NavItem');
var SelectableContext = require('@restart/ui/SelectableContext');
var BaseDropdown = require('@restart/ui/Dropdown');
var useEventCallback = require('@restart/hooks/useEventCallback');
var DropdownItem$1 = require('@restart/ui/DropdownItem');
var DropdownMenu$1 = require('@restart/ui/DropdownMenu');
var useIsomorphicEffect = require('@restart/hooks/useIsomorphicEffect');
var useMergedRefs = require('@restart/hooks/useMergedRefs');
var warning = require('warning');
var types = require('../utils/types');
var useWrappedRefWithWarning = require('../utils/useWrappedRefWithWarning');
var DropdownToggle$1 = require('@restart/ui/DropdownToggle');
var DropdownContext$1 = require('@restart/ui/DropdownContext');
var Button$1 = require('@restart/ui/Button');
var TabContext = require('@restart/ui/TabContext');
var TabPanel = require('@restart/ui/TabPanel');
var Transition = require('react-transition-group/Transition');
var transitionEndListener = require('../utils/transitionEndListener');
var triggerBrowserReflow = require('../utils/triggerBrowserReflow');
var TransitionWrapper = require('../utils/TransitionWrapper');
var getTabTransitionComponent = require('../utils/getTabTransitionComponent');
var ElementChildren = require('../utils/ElementChildren');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var BaseTabs__default = /*#__PURE__*/_interopDefaultLegacy(BaseTabs);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var BaseNav__default = /*#__PURE__*/_interopDefaultLegacy(BaseNav);
var createWithBsPrefix__default = /*#__PURE__*/_interopDefaultLegacy(createWithBsPrefix);
var Anchor__default = /*#__PURE__*/_interopDefaultLegacy(Anchor);
var SelectableContext__default = /*#__PURE__*/_interopDefaultLegacy(SelectableContext);
var BaseDropdown__default = /*#__PURE__*/_interopDefaultLegacy(BaseDropdown);
var useEventCallback__default = /*#__PURE__*/_interopDefaultLegacy(useEventCallback);
var useIsomorphicEffect__default = /*#__PURE__*/_interopDefaultLegacy(useIsomorphicEffect);
var useMergedRefs__default = /*#__PURE__*/_interopDefaultLegacy(useMergedRefs);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var useWrappedRefWithWarning__default = /*#__PURE__*/_interopDefaultLegacy(useWrappedRefWithWarning);
var DropdownContext__default = /*#__PURE__*/_interopDefaultLegacy(DropdownContext$1);
var TabContext__default = /*#__PURE__*/_interopDefaultLegacy(TabContext);
var transitionEndListener__default = /*#__PURE__*/_interopDefaultLegacy(transitionEndListener);
var triggerBrowserReflow__default = /*#__PURE__*/_interopDefaultLegacy(triggerBrowserReflow);
var TransitionWrapper__default = /*#__PURE__*/_interopDefaultLegacy(TransitionWrapper);
var getTabTransitionComponent__default = /*#__PURE__*/_interopDefaultLegacy(getTabTransitionComponent);

const ThemeContext = React__namespace.createContext({ prefixes: {} });
({
    prefixes: PropTypes__default["default"].object,
    dir: PropTypes__default["default"].string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = React.useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
// `sgds-${defaultPrefix}`
function useIsRTL() {
    const { dir } = React.useContext(ThemeContext);
    return dir === 'rtl';
}
const SGDSWrapper = React__namespace.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React__namespace.createElement(Tag, { ref: ref, ...props, className: classNames__default["default"](props.className, 'sgds') });
});

const context$1 = React__namespace.createContext(null);
context$1.displayName = 'NavbarContext';

const NavItem = createWithBsPrefix__default["default"]('nav-item', {
    Component: "li",
    displayName: 'NavItem'
});

const propTypes$c = {
    /**
     * @default 'nav-link'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * The active state of the NavItem item.
     */
    active: PropTypes__default["default"].bool,
    /**
     * The disabled state of the NavItem item.
     */
    disabled: PropTypes__default["default"].bool,
    /**
     * The ARIA role for the `NavLink`, In the context of a 'tablist' parent Nav,
     * the role defaults to 'tab'
     * */
    role: PropTypes__default["default"].string,
    /** The HTML href attribute for the `NavLink` */
    href: PropTypes__default["default"].string,
    /**
     * Uniquely identifies the `NavItem` amongst its siblings,
     * used to determine and control the active state of the parent `Nav`
     */
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /** @default 'a' */
    as: PropTypes__default["default"].elementType,
};
const defaultProps$5 = {
    disabled: false,
};
const NavLink = React__namespace.forwardRef(({ bsPrefix, className, as: Component = Anchor__default["default"], active, eventKey, 
// onClick,
...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'nav-link');
    const [navItemProps, meta] = NavItem$1.useNavItem({
        key: SelectableContext.makeEventKey(eventKey, props.href),
        active,
        ...props,
    });
    return (React__namespace.createElement(Component, { ...props, ...navItemProps, ref: ref, className: classNames__default["default"](className, bsPrefix, props.disabled && 'disabled', meta.isActive && 'active'), "aria-current": meta.isActive ? "page" : "false" }));
});
NavLink.displayName = 'NavLink';
NavLink.propTypes = propTypes$c;
NavLink.defaultProps = defaultProps$5;

const DropdownContext = React__namespace.createContext({});
DropdownContext.displayName = 'DropdownContext';

const propTypes$b = {
    /** @default 'dropdown-item' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * Highlight the menu item as active.
     */
    active: PropTypes__default["default"].bool,
    /**
     * Disable the menu item, making it unselectable.
     */
    disabled: PropTypes__default["default"].bool,
    /**
     * Value passed to the `onSelect` handler, useful for identifying the selected menu item.
     */
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * HTML `href` attribute corresponding to `a.href`.
     */
    href: PropTypes__default["default"].string,
    /**
     * Callback fired when the menu item is clicked.
     */
    onClick: PropTypes__default["default"].func,
    as: PropTypes__default["default"].elementType,
};
const DropdownItem = React__namespace.forwardRef(({ bsPrefix, className, eventKey, disabled = false, onClick, active, as: Component = Anchor__default["default"], ...props }, ref) => {
    const prefix = 'dropdown-item';
    const [dropdownItemProps, meta] = DropdownItem$1.useDropdownItem({
        key: eventKey,
        href: props.href,
        disabled,
        onClick,
        active,
    });
    return (React__namespace.createElement("li", null,
        React__namespace.createElement(Component, { ...props, ...dropdownItemProps, ref: ref, className: classNames__default["default"](className, prefix, meta.isActive && 'active', disabled && 'disabled') })));
});
DropdownItem.displayName = 'DropdownItem';
DropdownItem.propTypes = propTypes$b;

const context = React__namespace.createContext(null);
context.displayName = 'InputGroupContext';

const propTypes$a = {
    /**
     * @default 'dropdown-menu'
     */
    bsPrefix: PropTypes__default["default"].string,
    /** Controls the visibility of the Dropdown menu  */
    show: PropTypes__default["default"].bool,
    /** Whether to render the dropdown menu in the DOM before the first time it is shown */
    renderOnMount: PropTypes__default["default"].bool,
    /** Have the dropdown switch to it's opposite placement when necessary to stay on screen. */
    flip: PropTypes__default["default"].bool,
    /**
     * Aligns the dropdown menu to the specified side of the container. You can also align
     * the menu responsively for breakpoints starting at `sm` and up. The alignment
     * direction will affect the specified breakpoint or larger.
     *
     * *Note: Using responsive alignment will disable Popper usage for positioning.*
     *
     * @type {"start"|"end"|{ sm: "start"|"end" }|{ md: "start"|"end" }|{ lg: "start"|"end" }|{ xl: "start"|"end"}|{ xxl: "start"|"end"} }
     */
    align: types.alignPropType,
    onSelect: PropTypes__default["default"].func,
    /**
     * Which event when fired outside the component will cause it to be closed
     *
     * *Note: For custom dropdown components, you will have to pass the
     * `rootCloseEvent` to `<RootCloseWrapper>` in your custom dropdown menu
     * component ([similarly to how it is implemented in `<Dropdown.Menu>`](https://github.com/react-bootstrap/react-bootstrap/blob/v0.31.5/src/DropdownMenu.js#L115-L119)).*
     */
    rootCloseEvent: PropTypes__default["default"].oneOf(['click', 'mousedown']),
    /**
     * Control the rendering of the DropdownMenu. All non-menu props
     * (listed here) are passed through to the `as` Component.
     *
     * If providing a custom, non DOM, component. the `show`, `close` and `align` props
     * are also injected and should be handled appropriately.
     */
    as: PropTypes__default["default"].elementType,
    /**
     * A set of popper options and props passed directly to Popper.
     */
    popperConfig: PropTypes__default["default"].object,
    /**
     * Menu color variant.
     *
     * Omitting this will use the default light color.
     */
    variant: PropTypes__default["default"].string,
    /** @private Provides context when used in NavDropdown component  */
    isNav: PropTypes__default["default"].bool,
};
const defaultProps$4 = {
    flip: true,
};
function getDropdownMenuPlacement(alignEnd, dropDirection, isRTL) {
    const topStart = isRTL ? 'top-end' : 'top-start';
    const topEnd = isRTL ? 'top-start' : 'top-end';
    const bottomStart = isRTL ? 'bottom-end' : 'bottom-start';
    const bottomEnd = isRTL ? 'bottom-start' : 'bottom-end';
    // const leftStart = isRTL ? 'right-start' : 'left-start';
    // const leftEnd = isRTL ? 'right-end' : 'left-end';
    const rightStart = isRTL ? 'left-start' : 'right-start';
    const rightEnd = isRTL ? 'left-end' : 'right-end';
    let placement = alignEnd ? bottomEnd : bottomStart;
    if (dropDirection === 'up')
        placement = alignEnd ? topEnd : topStart;
    else if (dropDirection === 'end')
        placement = alignEnd ? rightEnd : rightStart;
    // else if (dropDirection === 'start')
    //   placement = alignEnd ? leftEnd : leftStart;
    return placement;
}
const DropdownMenu = React__namespace.forwardRef(({ bsPrefix, className, align, rootCloseEvent, flip, show: showProps, renderOnMount, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = 'ul', popperConfig, variant, isNav, ...props }, ref) => {
    let alignEnd = false;
    const isNavbar = React.useContext(context$1);
    const prefix = 'dropdown-menu';
    const { align: contextAlign, drop, isRTL } = React.useContext(DropdownContext);
    align = align || contextAlign;
    const isInputGroup = React.useContext(context);
    const alignClasses = [];
    if (align) {
        if (typeof align === 'object') {
            const keys = Object.keys(align);
            warning__default["default"](keys.length === 1, 'There should only be 1 breakpoint when passing an object to `align`');
            if (keys.length) {
                const brkPoint = keys[0];
                const direction = align[brkPoint];
                // .dropdown-menu-end is required for responsively aligning
                // left in addition to align left classes.
                alignEnd = direction === 'start';
                alignClasses.push(`${prefix}-${brkPoint}-${direction}`);
            }
        }
        else if (align === 'end') {
            alignEnd = true;
        }
    }
    const placement = getDropdownMenuPlacement(alignEnd, drop, isRTL);
    const [menuProps, { hasShown, popper, show, toggle }] = DropdownMenu$1.useDropdownMenu({
        flip,
        rootCloseEvent,
        show: showProps,
        usePopper: !isNavbar && alignClasses.length === 0,
        offset: !isNav ? [0, 10] : undefined,
        popperConfig,
        placement,
    });
    menuProps.ref = useMergedRefs__default["default"](useWrappedRefWithWarning__default["default"](ref, 'DropdownMenu'), menuProps.ref);
    useIsomorphicEffect__default["default"](() => {
        // Popper's initial position for the menu is incorrect when
        // renderOnMount=true. Need to call update() to correct it.
        if (show)
            popper?.update();
    }, [show]);
    if (!hasShown && !renderOnMount && !isInputGroup)
        return null;
    // For custom components provide additional, non-DOM, props;
    if (typeof Component !== 'string') {
        menuProps.show = show;
        menuProps.close = () => toggle?.(false);
        menuProps.align = align;
    }
    let style = props.style;
    if (popper?.placement) {
        // we don't need the default popper style,
        // menus are display: none when not shown.
        style = { ...props.style, ...menuProps.style };
        //@ts-ignore
        props['x-placement'] = popper.placement;
    }
    return (React__namespace.createElement(Component, { ...props, ...menuProps, style: style, ...((alignClasses.length || isNavbar) && {
            'data-bs-popper': 'static',
        }), className: classNames__default["default"](className, prefix, show && 'show', alignEnd && `${prefix}-end`, variant && `${prefix}-${variant}`, ...alignClasses) }));
});
DropdownMenu.displayName = 'DropdownMenu';
DropdownMenu.propTypes = propTypes$a;
DropdownMenu.defaultProps = defaultProps$4;

const propTypes$9 = {
    /**
     * @default 'btn'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * One or more button variant combinations
     *
     * buttons may be one of a variety of visual variants such as:
     *
     * `'primary', 'secondary', 'success', 'danger', 'warning', 'info', 'dark', 'light', 'link'`
     *
     * as well as "outline" versions (prefixed by 'outline-*')
     *
     * `'outline-primary', 'outline-secondary', 'outline-success', 'outline-danger', 'outline-warning', 'outline-info', 'outline-dark', 'outline-light'`
     */
    variant: PropTypes__default["default"].string,
    /**
     * Specifies a large or small button.
     *
     * @type ('sm'|'lg')
     */
    size: PropTypes__default["default"].string,
    /** Manually set the visual state of the button to `:active` */
    active: PropTypes__default["default"].bool,
    /**
     * Disables the Button, preventing mouse events,
     * even if the underlying component is an `<a>` element
     */
    disabled: PropTypes__default["default"].bool,
    /** Providing a `href` will render an `<a>` element, _styled_ as a button. */
    href: PropTypes__default["default"].string,
    /**
     * Defines HTML button type attribute.
     *
     * @default 'button'
     */
    type: PropTypes__default["default"].oneOf(['button', 'reset', 'submit', null]),
    as: PropTypes__default["default"].elementType,
};
const defaultProps$3 = {
    variant: 'primary',
    active: false,
    disabled: false,
};
const Button = React__namespace.forwardRef(({ as, bsPrefix, variant, size, active, className, ...props }, ref) => {
    const prefix = useBootstrapPrefix(bsPrefix, 'btn');
    const [buttonProps, { tagName }] = Button$1.useButtonProps({
        tagName: as,
        ...props,
    });
    const Component = tagName;
    return (React__namespace.createElement(SGDSWrapper, { as: Component, ...props, ...buttonProps, ref: ref, className: classNames__default["default"](className, prefix, active && 'active', variant && `${prefix}-${variant}`, size && `${prefix}-${size}`, props.href && props.disabled && 'disabled') }));
});
Button.displayName = 'Button';
Button.propTypes = propTypes$9;
Button.defaultProps = defaultProps$3;

const propTypes$8 = {
    /**
     * @default 'dropdown-toggle'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * An html id attribute, necessary for assistive technologies, such as screen readers.
     * @type {string|number}
     */
    id: PropTypes__default["default"].string,
    split: PropTypes__default["default"].bool,
    as: PropTypes__default["default"].elementType,
    /**
     * to passthrough to the underlying button or whatever from DropdownButton
     * @private
     */
    childBsPrefix: PropTypes__default["default"].string,
};
const DropdownToggle = React__namespace.forwardRef(({ bsPrefix, split, className, childBsPrefix, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = Button, ...props }, ref) => {
    const prefix = 'dropdown-toggle';
    const dropdownContext = React.useContext(DropdownContext__default["default"]);
    const isInputGroup = React.useContext(context);
    if (childBsPrefix !== undefined) {
        props.bsPrefix = childBsPrefix;
    }
    const [toggleProps] = DropdownToggle$1.useDropdownToggle();
    toggleProps.ref = useMergedRefs__default["default"](toggleProps.ref, useWrappedRefWithWarning__default["default"](ref, 'DropdownToggle'));
    // This intentionally forwards size and variant (if set) to the
    // underlying component, to allow it to render size and style variants.
    return (React__namespace.createElement(Component, { className: classNames__default["default"](className, prefix, split && `${prefix}-split`, !!isInputGroup && dropdownContext?.show && 'show'), variant: "outline-secondary", ...toggleProps, ...props, "aria-haspopup": "menu" }));
});
DropdownToggle.displayName = 'DropdownToggle';
DropdownToggle.propTypes = propTypes$8;

const DropdownHeader = createWithBsPrefix__default["default"]('dropdown-header', {
    defaultProps: { role: 'heading' },
});
const DropdownDivider = createWithBsPrefix__default["default"]('dropdown-divider', {
    Component: 'hr',
    defaultProps: { role: 'separator' },
});
const DropdownItemText = createWithBsPrefix__default["default"]('dropdown-item-text', {
    Component: 'span',
});

const propTypes$7 = {
    /** @default 'dropdown' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * Determines the direction and location of the Menu in relation to it's Toggle.
     *
     * @type {"end"|"up"|"down" }
     */
    drop: PropTypes__default["default"].oneOf(['up', 'end', 'down']),
    as: PropTypes__default["default"].elementType,
    align: types.alignDirection,
    /**
     * Whether or not the Dropdown is visible.
     *
     * @controllable onToggle
     */
    show: PropTypes__default["default"].bool,
    /**
     * Allow Dropdown to flip in case of an overlapping on the reference element. For more information refer to
     * Popper.js's flip [docs](https://popper.js.org/docs/v2/modifiers/flip/).
     *
     */
    flip: PropTypes__default["default"].bool,
    /**
     * A callback fired when the Dropdown wishes to change visibility. Called with the requested
     * `show` value, the DOM event, and the source that fired it: `'click'`,`'keydown'`,`'rootClose'`, or `'select'`.
     *
     * ```js
     * function(
     *   isOpen: boolean,
     *   event: SyntheticEvent,
     *   metadata: {
     *     source: 'select' | 'click' | 'rootClose' | 'keydown'
     *   }
     * ): void
     * ```
     *
     * @controllable show
     */
    onToggle: PropTypes__default["default"].func,
    /**
     * A callback fired when a menu item is selected.
     *
     * ```js
     * (eventKey: any, event: Object) => any
     * ```
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * Controls the focus behavior for when the Dropdown is opened. Set to
     * `true` to always focus the first menu item, `keyboard` to focus only when
     * navigating via the keyboard, or `false` to disable completely
     *
     * The Default behavior is `false` **unless** the Menu has a `role="menu"`
     * where it will default to `keyboard` to match the recommended [ARIA Authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton).
     */
    focusFirstItemOnShow: PropTypes__default["default"].oneOf([false, true, 'keyboard']),
    /** @private */
    navbar: PropTypes__default["default"].bool,
    /**
     * Controls the auto close behaviour of the dropdown when clicking outside of
     * the button or the list.
     */
    autoClose: PropTypes__default["default"].oneOf([true, 'outside', 'inside', false]),
};
const defaultProps$2 = {
    navbar: false,
    align: 'start',
    autoClose: true,
    focusFirstItemOnShow: 'keyboard',
};
const Dropdown = React__namespace.forwardRef((pProps, ref) => {
    const { bsPrefix, drop, show, className, align, onSelect, onToggle, focusFirstItemOnShow, 
    // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
    as: Component = 'div', navbar: _4, autoClose, ...props } = uncontrollable.useUncontrolled(pProps, { show: 'onToggle' });
    const isInputGroup = React.useContext(context);
    const prefix = useBootstrapPrefix(bsPrefix, 'dropdown');
    const isRTL = useIsRTL();
    const isClosingPermitted = (source) => {
        // autoClose=false only permits close on button click
        if (autoClose === false)
            return source === 'click';
        // autoClose=inside doesn't permit close on rootClose
        if (autoClose === 'inside')
            return source !== 'rootClose';
        // autoClose=outside doesn't permit close on select
        if (autoClose === 'outside')
            return source !== 'select';
        return true;
    };
    const handleToggle = useEventCallback__default["default"]((nextShow, meta) => {
        if (meta.originalEvent.currentTarget === document &&
            (meta.source !== 'keydown' ||
                meta.originalEvent.key === 'Escape'))
            meta.source = 'rootClose';
        if (isClosingPermitted(meta.source))
            onToggle?.(nextShow, meta);
    });
    const alignEnd = align === 'end';
    const placement = getDropdownMenuPlacement(alignEnd, drop, isRTL);
    const contextValue = React.useMemo(() => ({
        align,
        drop,
        isRTL,
    }), [align, drop, isRTL]);
    return (React__namespace.createElement(DropdownContext.Provider, { value: contextValue },
        React__namespace.createElement(BaseDropdown__default["default"], { placement: placement, show: show, onSelect: onSelect, onToggle: handleToggle, focusFirstItemOnShow: focusFirstItemOnShow, itemSelector: `.${prefix}-item:not(.disabled):not(:disabled)` }, isInputGroup ? (props.children) : (React__namespace.createElement(SGDSWrapper, { as: Component, ...props, ref: ref, className: classNames__default["default"](className, show && 'show', prefix, drop === 'up' && 'dropup', drop === 'end' && 'dropend') })))));
});
Dropdown.displayName = 'Dropdown';
Dropdown.propTypes = propTypes$7;
Dropdown.defaultProps = defaultProps$2;
var Dropdown$1 = Object.assign(Dropdown, {
    Toggle: DropdownToggle,
    Menu: DropdownMenu,
    Item: DropdownItem,
    ItemText: DropdownItemText,
    Divider: DropdownDivider,
    Header: DropdownHeader,
});

const NavContext = React__namespace.createContext(null);
NavContext.displayName = 'NavContext';

const propTypes$6 = {
    /**
     * An html id attribute for the Toggle button, necessary for assistive technologies, such as screen readers.
     * @type {string}
     */
    id: PropTypes__default["default"].string,
    /** An `onClick` handler passed to the Toggle component */
    onClick: PropTypes__default["default"].func,
    /** The content of the non-toggle Button.  */
    title: PropTypes__default["default"].node.isRequired,
    /** Disables the toggle NavLink  */
    disabled: PropTypes__default["default"].bool,
    /** Style the toggle NavLink as active  */
    active: PropTypes__default["default"].bool,
    /** An ARIA accessible role applied to the Menu component. When set to 'menu', The dropdown */
    menuRole: PropTypes__default["default"].string,
    /** Whether to render the dropdown menu in the DOM before the first time it is shown */
    renderMenuOnMount: PropTypes__default["default"].bool,
    /**
     *  Which event when fired outside the component will cause it to be closed.
     *
     * _see [DropdownMenu](#menu-props) for more details_
     */
    rootCloseEvent: PropTypes__default["default"].string,
    /**
     * Menu color variant.
     *
     * Omitting this will use the default light color.
     */
    menuVariant: PropTypes__default["default"].oneOf(['dark']),
    /** @ignore */
    bsPrefix: PropTypes__default["default"].string,
    /** When true, applies mega menu stylings */
    isMegaMenu: PropTypes__default["default"].bool,
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
};
const NavDropdown = React__namespace.forwardRef(({ id, title, children, bsPrefix, className, rootCloseEvent, menuRole, disabled, active, renderMenuOnMount, menuVariant, isMegaMenu, eventKey, align, ...props }, ref) => {
    /* NavItem has no additional logic, it's purely presentational. Can set nav item class here to support "as" */
    const navContext = React.useContext(NavContext);
    const navItemPrefix = useBootstrapPrefix(undefined, 'nav-item');
    const dropDownClass = classNames__default["default"](className, navItemPrefix, isMegaMenu ? 'has-megamenu' : undefined);
    return (React__namespace.createElement(Dropdown$1, { ref: ref, as: "li", ...props, className: dropDownClass, align: align },
        React__namespace.createElement(Dropdown$1.Toggle, { id: id, eventKey: null, active: active || navContext?.activeKey === eventKey, disabled: disabled, childBsPrefix: bsPrefix, as: NavLink },
            title,
            React__namespace.createElement("i", { className: "bi bi-chevron-down" })),
        React__namespace.createElement(Dropdown$1.Menu, { role: menuRole, renderOnMount: renderMenuOnMount, rootCloseEvent: rootCloseEvent, variant: menuVariant, isNav: true, align: align, as: isMegaMenu ? 'div' : undefined, className: isMegaMenu ? 'mega-menu' : undefined }, children)));
});
NavDropdown.displayName = 'NavDropdown';
NavDropdown.propTypes = propTypes$6;
var NavDropdown$1 = Object.assign(NavDropdown, {
    Item: Dropdown$1.Item,
    ItemText: Dropdown$1.ItemText,
    Divider: Dropdown$1.Divider,
    Header: Dropdown$1.Header,
});

const propTypes$5 = {
    /**
     * @default 'nav'
     */
    bsPrefix: PropTypes__default["default"].string,
    /** @private */
    // navbarBsPrefix: PropTypes.string,
    /** @private */
    // cardHeaderBsPrefix: PropTypes.string,
    /**
     * The visual variant of the nav items.
     *
     * @type {('tabs-basic-toggle' | 'tabs-info-toggle')}
     */
    variant: PropTypes__default["default"].string,
    /**
     * The default active key that is selected on start.
     */
    defaultActiveKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * Marks the NavItem with a matching `eventKey` (or `href` if present) as active.
     */
    activeKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * A callback fired when a NavItem is selected.
     *
     * ```js
     * function (
     *  Any eventKey,
     *  SyntheticEvent event?
     * )
     * ```
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * ARIA role for the Nav, in the context of a TabContainer, the default will
     * be set to "tablist", but can be overridden by the Nav when set explicitly.
     *
     * When the role is "tablist", NavLink focus is managed according to
     * the ARIA authoring practices for tabs:
     * https://www.w3.org/TR/2013/WD-wai-aria-practices-20130307/#tabpanel
     */
    role: PropTypes__default["default"].string,
    /**
     * Apply styling an alignment for use in a Navbar. This prop will be set
     * automatically when the Nav is used inside a Navbar.
     */
    // navbar: PropTypes.bool,
    /**
     * Enable vertical scrolling within the toggleable contents of a collapsed Navbar.
     */
    navbarScroll: PropTypes__default["default"].bool,
    as: PropTypes__default["default"].elementType,
    /** @private */
    onKeyDown: PropTypes__default["default"].func,
};
const Nav = React__namespace.forwardRef((uncontrolledProps, ref) => {
    const { as = 'ul', bsPrefix: initialBsPrefix, navbarScroll, className, activeKey, ...props } = uncontrollable.useUncontrolled(uncontrolledProps, { activeKey: 'onSelect' });
    const bsPrefix = useBootstrapPrefix(initialBsPrefix, 'nav');
    const contextValue = React.useMemo(() => ({
        activeKey,
    }), [activeKey]);
    const navbarContext = React.useContext(context$1);
    return (React__namespace.createElement(NavContext.Provider, { value: contextValue },
        React__namespace.createElement(BaseNav__default["default"], { as: as, ref: ref, activeKey: activeKey, className: classNames__default["default"](className, {
                [bsPrefix]: !navbarContext,
                [`${navbarContext?.bsPrefix}-nav`]: navbarContext,
                [`${navbarContext?.bsPrefix}-nav-scroll`]: navbarContext && navbarScroll,
                ['sgds']: !navbarContext,
                ['nav-tabs']: !navbarContext,
            }), ...props })));
});
Nav.displayName = 'Nav';
Nav.propTypes = propTypes$5;
var Nav$1 = Object.assign(Nav, {
    Item: NavItem,
    Link: NavLink,
    Dropdown: NavDropdown$1,
});

const TabContent = createWithBsPrefix__default["default"]('sgds tab-content');

const propTypes$4 = {
    /**
     * Show the component; triggers the fade in or fade out animation
     */
    in: PropTypes__default["default"].bool,
    /**
     * Wait until the first "enter" transition to mount the component (add it to the DOM)
     */
    mountOnEnter: PropTypes__default["default"].bool,
    /**
     * Unmount the component (remove it from the DOM) when it is faded out
     */
    unmountOnExit: PropTypes__default["default"].bool,
    /**
     * Run the fade in animation when the component mounts, if it is initially
     * shown
     */
    appear: PropTypes__default["default"].bool,
    /**
     * Duration of the fade animation in milliseconds, to ensure that finishing
     * callbacks are fired even if the original browser transition end events are
     * canceled
     */
    timeout: PropTypes__default["default"].number,
    /**
     * Callback fired before the component fades in
     */
    onEnter: PropTypes__default["default"].func,
    /**
     * Callback fired after the component starts to fade in
     */
    onEntering: PropTypes__default["default"].func,
    /**
     * Callback fired after the has component faded in
     */
    onEntered: PropTypes__default["default"].func,
    /**
     * Callback fired before the component fades out
     */
    onExit: PropTypes__default["default"].func,
    /**
     * Callback fired after the component starts to fade out
     */
    onExiting: PropTypes__default["default"].func,
    /**
     * Callback fired after the component has faded out
     */
    onExited: PropTypes__default["default"].func,
    /**
     * You must provide a single JSX child element to this component and that element cannot be a \<React.Fragment\>
     */
    children: PropTypes__default["default"].element.isRequired,
    /**
     * Applies additional specified classes during the transition. Takes an object
     * where the keys correspond to the Transition status
     */
    transitionClasses: PropTypes__default["default"].object,
};
const defaultProps$1 = {
    in: false,
    timeout: 300,
    mountOnEnter: false,
    unmountOnExit: false,
    appear: false,
};
const fadeStyles = {
    [Transition.ENTERING]: 'show',
    [Transition.ENTERED]: 'show',
};
const Fade = React__namespace.forwardRef(({ className, children, transitionClasses = {}, ...props }, ref) => {
    const handleEnter = React.useCallback((node, isAppearing) => {
        triggerBrowserReflow__default["default"](node);
        props.onEnter?.(node, isAppearing);
    }, [props]);
    return (React__namespace.createElement(TransitionWrapper__default["default"], { ref: ref, addEndListener: transitionEndListener__default["default"], ...props, onEnter: handleEnter, childRef: children.ref }, (status, innerProps) => React__namespace.cloneElement(children, {
        ...innerProps,
        className: classNames__default["default"]('fade', className, children.props.className, 
        //@ts-ignore
        fadeStyles[status], transitionClasses[status]),
    })));
});
Fade.propTypes = propTypes$4;
Fade.defaultProps = defaultProps$1;
Fade.displayName = 'Fade';

const propTypes$3 = {
    /**
     * @default 'tab-pane'
     */
    bsPrefix: PropTypes__default["default"].string,
    as: PropTypes__default["default"].elementType,
    /**
     * A key that associates the `TabPane` with it's controlling `NavLink`.
     */
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * Toggles the active state of the TabPane, this is generally controlled by a
     * TabContainer.
     */
    active: PropTypes__default["default"].bool,
    /**
     * Use animation when showing or hiding `<TabPane>`s. Defaults to `<Fade>`
     * animation, else use `false` to disable or a react-transition-group
     * `<Transition/>` component.
     */
    transition: PropTypes__default["default"].oneOfType([PropTypes__default["default"].bool, PropTypes__default["default"].elementType]),
    /**
     * Transition onEnter callback when animation is not `false`
     */
    onEnter: PropTypes__default["default"].func,
    /**
     * Transition onEntering callback when animation is not `false`
     */
    onEntering: PropTypes__default["default"].func,
    /**
     * Transition onEntered callback when animation is not `false`
     */
    onEntered: PropTypes__default["default"].func,
    /**
     * Transition onExit callback when animation is not `false`
     */
    onExit: PropTypes__default["default"].func,
    /**
     * Transition onExiting callback when animation is not `false`
     */
    onExiting: PropTypes__default["default"].func,
    /**
     * Transition onExited callback when animation is not `false`
     */
    onExited: PropTypes__default["default"].func,
    /**
     * Wait until the first "enter" transition to mount the tab (add it to the DOM)
     */
    mountOnEnter: PropTypes__default["default"].bool,
    /**
     * Unmount the tab (remove it from the DOM) when it is no longer visible
     */
    unmountOnExit: PropTypes__default["default"].bool,
    /** @ignore * */
    id: PropTypes__default["default"].string,
    /** @ignore * */
    'aria-labelledby': PropTypes__default["default"].string,
};
const TabPane = React__namespace.forwardRef(({ bsPrefix, transition, ...props }, ref) => {
    const [{ className, 
    // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
    as: Component = 'div', ...rest }, { isActive, onEnter, onEntering, onEntered, onExit, onExiting, onExited, mountOnEnter, unmountOnExit, transition: Transition = Fade, },] = TabPanel.useTabPanel({
        ...props,
        transition: getTabTransitionComponent__default["default"](transition),
    });
    const { mdxType, originalType, contentLeft, contentRight, contentBottom, variant, ...attrs } = rest;
    const prefix = useBootstrapPrefix(bsPrefix, 'tab-pane');
    // We provide an empty the TabContext so `<Nav>`s in `<TabPanel>`s don't
    // conflict with the top level one.
    return (React__namespace.createElement(TabContext__default["default"].Provider, { value: null },
        React__namespace.createElement(SelectableContext__default["default"].Provider, { value: null },
            React__namespace.createElement(Transition, { in: isActive, onEnter: onEnter, onEntering: onEntering, onEntered: onEntered, onExit: onExit, onExiting: onExiting, onExited: onExited, mountOnEnter: mountOnEnter, unmountOnExit: unmountOnExit },
                React__namespace.createElement(Component, { ...attrs, ref: ref, className: classNames__default["default"](className, prefix, isActive && 'active') })))));
});
TabPane.displayName = 'TabPane';
TabPane.propTypes = propTypes$3;

const propTypes$2 = {
    /**
     * Mark the Tab with a matching `eventKey` as active.
     *
     * @controllable onSelect
     */
    activeKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /** The default active key that is selected on start */
    defaultActiveKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * Navigation style
     *
     * @type {('tabs-basic-toggle'| 'tabs-info-toggle')}
     */
    variant: PropTypes__default["default"].string,
    /**
     * Sets a default animation strategy for all children `<TabPane>`s.<tbcont
     *
     * Defaults to `<Fade>` animation, else use `false` to disable or a
     * react-transition-group `<Transition/>` component.
     *
     * @type {Transition | false}
     * @default {Fade}
     */
    transition: PropTypes__default["default"].oneOfType([
        PropTypes__default["default"].oneOf([false]),
        PropTypes__default["default"].elementType,
    ]),
    /**
     * HTML id attribute, required if no `generateChildId` prop
     * is specified.
     *
     * @type {string}
     */
    id: PropTypes__default["default"].string,
    /**
     * Callback fired when a Tab is selected.
     *
     * ```js
     * function (
     *   Any eventKey,
     *   SyntheticEvent event?
     * )
     * ```
     *
     * @controllable activeKey
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * Wait until the first "enter" transition to mount tabs (add them to the DOM)
     */
    mountOnEnter: PropTypes__default["default"].bool,
    /**
     * Unmount tabs (remove it from the DOM) when it is no longer visible
     */
    unmountOnExit: PropTypes__default["default"].bool,
};
const defaultProps = {
    // variant: 'tabs',
    mountOnEnter: false,
    unmountOnExit: false,
};
function getDefaultActiveKey(children) {
    let defaultActiveKey;
    ElementChildren.forEach(children, (child) => {
        if (defaultActiveKey == null) {
            defaultActiveKey = child.props.eventKey;
        }
    });
    return defaultActiveKey;
}
function renderTab(variant) {
    return (child) => {
        const { title, eventKey, disabled, tabClassName, tabAttrs, id, contentLeft, contentRight, contentBottom, } = child.props;
        if (title == null) {
            return null;
        }
        const tabsBasic = (React__namespace.createElement(React__namespace.Fragment, null,
            contentLeft,
            title,
            contentRight));
        const tabsInfo = (React__namespace.createElement(React__namespace.Fragment, null,
            React__namespace.createElement("div", { className: `tabs-info-label ${contentLeft ? 'has-icon' : ''}` },
                contentLeft,
                title),
            React__namespace.createElement("div", { className: "tabs-info-count" }, contentBottom)));
        return (React__namespace.createElement(NavItem, { as: "li", role: "presentation" },
            React__namespace.createElement(NavLink, { as: "button", type: "button", eventKey: eventKey, disabled: disabled, id: id, className: tabClassName, ...tabAttrs }, variant === 'tabs-info-toggle' ? tabsInfo : tabsBasic)));
    };
}
const Tabs = (props) => {
    const { id, variant, onSelect, transition, mountOnEnter, unmountOnExit, children, activeKey = getDefaultActiveKey(children), ...controlledProps } = uncontrollable.useUncontrolled(props, {
        activeKey: 'onSelect',
    });
    return (React__namespace.createElement(BaseTabs__default["default"], { id: id, activeKey: activeKey, onSelect: onSelect, transition: getTabTransitionComponent__default["default"](transition), mountOnEnter: mountOnEnter, unmountOnExit: unmountOnExit },
        React__namespace.createElement(Nav$1, { ...controlledProps, role: "tablist", as: "ul", variant: props.variant }, ElementChildren.map(children, renderTab(variant))),
        React__namespace.createElement(TabContent, null, ElementChildren.map(children, (child) => {
            const childProps = { ...child.props };
            delete childProps.title;
            delete childProps.disabled;
            delete childProps.tabClassName;
            delete childProps.tabAttrs;
            return React__namespace.createElement(TabPane, { ...childProps });
        }))));
};
Tabs.propTypes = propTypes$2;
Tabs.defaultProps = defaultProps;
Tabs.displayName = 'Tabs';

const propTypes$1 = {
    /**
     * HTML id attribute, required if no `generateChildId` prop
     * is specified.
     *
     * @type {string}
     */
    id: PropTypes__default["default"].string,
    /**
     * Sets a default animation strategy for all children `<TabPane>`s.
     * Defaults to `<Fade>` animation; else, use `false` to disable, or a
     * custom react-transition-group `<Transition/>` component.
     *
     * @type {{Transition | false}}
     * @default {Fade}
     */
    transition: PropTypes__default["default"].oneOfType([
        PropTypes__default["default"].oneOf([false]),
        PropTypes__default["default"].elementType,
    ]),
    /**
     * Wait until the first "enter" transition to mount tabs (add them to the DOM)
     */
    mountOnEnter: PropTypes__default["default"].bool,
    /**
     * Unmount tabs (remove it from the DOM) when they are no longer visible
     */
    unmountOnExit: PropTypes__default["default"].bool,
    /**
     * A function that takes an `eventKey` and `type` and returns a unique id for
     * child tab `<NavItem>`s and `<TabPane>`s. The function _must_ be a pure
     * function, meaning it should always return the _same_ id for the same set
     * of inputs. The default value requires that an `id` to be set for the
     * `<TabContainer>`.
     *
     * The `type` argument will either be `"tab"` or `"pane"`.
     *
     * @defaultValue (eventKey, type) => `${props.id}-${type}-${eventKey}`
     */
    generateChildId: PropTypes__default["default"].func,
    /**
     * A callback fired when a tab is selected.
     *
     * @controllable activeKey
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * The `eventKey` of the currently active tab.
     *
     * @controllable onSelect
     */
    activeKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
};
const TabContainer = ({ transition, ...props }) => (React__namespace.createElement(BaseTabs__default["default"], { ...props, transition: getTabTransitionComponent__default["default"](transition) }));
TabContainer.propTypes = propTypes$1;
TabContainer.displayName = 'TabContainer';

/* eslint-disable react/no-unused-prop-types */
const propTypes = {
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * Content for the tab title.
     */
    title: PropTypes__default["default"].node.isRequired,
    /**
     * The disabled state of the tab.
     */
    disabled: PropTypes__default["default"].bool,
    /**
     * Class to pass to the underlying nav link.
     */
    tabClassName: PropTypes__default["default"].string,
    /**
     * Object containing attributes to pass to underlying nav link.
     */
    tabAttrs: PropTypes__default["default"].object,
    /**
    * Content to the left of title.
    */
    contentLeft: PropTypes__default["default"].node,
    /**
    * Content to the right of title. Not applicable to tabs-info-toggle variant.
    */
    contentRight: PropTypes__default["default"].node,
    /**
     * Content to the bottom of title. Applicable to tabs-info-toggle variant only.
     */
    contentBottom: PropTypes__default["default"].node,
};
const Tab = () => {
    throw new Error('ReactBootstrap: The `Tab` component is not meant to be rendered! ' +
        "It's an abstract component that is only valid as a direct Child of the `Tabs` Component. " +
        'For custom tabs components use TabPane and TabsContainer directly');
};
Tab.propTypes = propTypes;
var Tab$1 = Object.assign(Tab, {
    Container: TabContainer,
    Content: TabContent,
    Pane: TabPane,
});

exports.Tab = Tab$1;
exports.TabContainer = TabContainer;
exports.TabContent = TabContent;
exports.TabPane = TabPane;
exports.Tabs = Tabs;
