// @govtechsg/sgds-react/Card/index.esm.js
'use client';
import classNames from 'classnames';
import * as React from 'react';
import { useContext, useMemo } from 'react';
import PropTypes from 'prop-types';
import createWithBsPrefix from '../utils/createWithBsPrefix';
import divWithClassName from '../utils/divWithClassName';

const ThemeContext = React.createContext({ prefixes: {} });
({
    prefixes: PropTypes.object,
    dir: PropTypes.string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
const SGDSWrapper = React.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React.createElement(Tag, { ref: ref, ...props, className: classNames(props.className, 'sgds') });
});

const propTypes$7 = {
    /**
     * @default 'card-img'
     */
    bsPrefix: PropTypes.string,
    /**
     * Defines image position inside
     * the card.
     *
     * @type {('top'|'bottom')}
     */
    variant: PropTypes.oneOf(['top', 'bottom']),
    as: PropTypes.elementType,
};
const CardImg = React.forwardRef(
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
({ bsPrefix, className, variant, as: Component = 'img', ...props }, ref) => {
    const prefix = useBootstrapPrefix(bsPrefix, 'card-img');
    return (React.createElement(Component, { ref: ref, className: classNames(variant ? `${prefix}-${variant}` : prefix, className), ...props }));
});
CardImg.displayName = 'CardImg';
CardImg.propTypes = propTypes$7;

const context = React.createContext(null);
context.displayName = 'CardHeaderContext';

const propTypes$6 = {
    /**
     * @default 'card-header'
     */
    bsPrefix: PropTypes.string,
    as: PropTypes.elementType,
};
const CardHeader = React.forwardRef(({ bsPrefix, className, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = 'div', ...props }, ref) => {
    const prefix = useBootstrapPrefix(bsPrefix, 'card-header');
    const contextValue = useMemo(() => ({
        cardHeaderBsPrefix: prefix,
    }), [prefix]);
    return (React.createElement(context.Provider, { value: contextValue },
        React.createElement(Component, { ref: ref, ...props, className: classNames(className, prefix) })));
});
CardHeader.displayName = 'CardHeader';
CardHeader.propTypes = propTypes$6;

const CardTitle = createWithBsPrefix('card-title', {
    Component: 'h3',
});
const DivStyledAsH6 = divWithClassName('h6');
const CardSubtitle = createWithBsPrefix('card-subtitle', {
    Component: DivStyledAsH6,
});
const CardLink = createWithBsPrefix('card-link', { Component: 'a' });
const CardStretchedLink = createWithBsPrefix('card-link stretched-link', {
    Component: 'a',
});
const CardText = createWithBsPrefix('card-text', { Component: 'p' });
const CardFooter = createWithBsPrefix('card-footer');
const CardImgOverlay = createWithBsPrefix('card-img-overlay');
const CardUnit = createWithBsPrefix('card-unit', { Component: 'div' });
const CardBody = createWithBsPrefix('card-body');

const propTypes$5 = {
    /**
     * @default 'card'
     */
    bsPrefix: PropTypes.string,
    /**
     * Sets card background
     *
     * @type {('primary'|'secondary'|'success'|'danger'|'warning'|'info'|'dark'|'light')}
     */
    bg: PropTypes.string,
    /**
     * Sets card text color
     *
     * @type {('primary'|'secondary'|'success'|'danger'|'warning'|'info'|'dark'|'light'|'white'|'muted')}
     */
    text: PropTypes.string,
    /**
     * Sets card border color
     *
     * @type {('primary'|'secondary'|'success'|'danger'|'warning'|'info'|'dark'|'light')}
     */
    border: PropTypes.string,
    as: PropTypes.elementType,
    variant: PropTypes.oneOf([
        'card-action',
        'card-action-quantity-toggle',
    ]),
};
const Card = React.forwardRef(({ bsPrefix, className, bg, text, border, children, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = 'div', ...props }, ref) => {
    const prefix = useBootstrapPrefix(bsPrefix, 'card');
    return (React.createElement(SGDSWrapper, { as: Component, ref: ref, ...props, className: classNames(className, prefix, bg && `bg-${bg}`, text && `text-${text}`, border && `border-${border}`) }, children));
});
Card.displayName = 'Card';
Card.propTypes = propTypes$5;
var Card$1 = Object.assign(Card, {
    Img: CardImg,
    Title: CardTitle,
    Subtitle: CardSubtitle,
    Body: CardBody,
    Link: CardLink,
    StretchedLink: CardStretchedLink,
    Text: CardText,
    Header: CardHeader,
    Footer: CardFooter,
    ImgOverlay: CardImgOverlay,
    Unit: CardUnit,
});

const propTypes$4 = {
    /**
     * Specify whether the feedback is for valid or invalid fields
     *
     * @type {('valid'|'invalid')}
     */
    type: PropTypes.string,
    as: PropTypes.elementType,
};
const Feedback = React.forwardRef(
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
({ as: Component = 'div', className, type = 'valid', ...props }, ref) => {
    return (React.createElement(Component, { ...props, ref: ref, className: classNames(className, `${type}-feedback`) }));
});
Feedback.displayName = 'Feedback';
Feedback.propTypes = propTypes$4;

const FormContext = React.createContext({});

const propTypes$3 = {
    /**
     * @default 'form-check-input'
     */
    bsPrefix: PropTypes.string,
    /**
     * The underlying HTML element to use when rendering the FormCheckInput.
     *
     * @type {('input'|elementType)}
     */
    as: PropTypes.elementType,
    /** A HTML id attribute, necessary for proper form accessibility. */
    id: PropTypes.string,
    /** The type of checkable. */
    type: PropTypes.oneOf(['radio', 'checkbox']).isRequired,
    /** Manually style the input as valid */
    isValid: PropTypes.bool,
    /** Manually style the input as invalid */
    isInvalid: PropTypes.bool,
};
const FormCheckInput = React.forwardRef(({ id, bsPrefix, className, type = 'checkbox', isValid = false, isInvalid = false, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = 'input', ...props }, ref) => {
    const { controlId } = useContext(FormContext);
    bsPrefix = useBootstrapPrefix(bsPrefix, 'form-check-input');
    return (React.createElement(Component, { ...props, ref: ref, type: type, id: id || controlId, className: classNames(className, bsPrefix, isValid && 'is-valid', isInvalid && 'is-invalid') }));
});
FormCheckInput.displayName = 'FormCheckInput';
FormCheckInput.propTypes = propTypes$3;

const propTypes$2 = {
    /**
     * @default 'form-check-label'
     */
    bsPrefix: PropTypes.string,
    /** The HTML for attribute for associating the label with an input */
    htmlFor: PropTypes.string,
};
const FormCheckLabel = React.forwardRef(({ bsPrefix, className, htmlFor, ...props }, ref) => {
    const { controlId } = useContext(FormContext);
    bsPrefix = useBootstrapPrefix(bsPrefix, 'form-check-label');
    return (React.createElement("label", { ...props, ref: ref, htmlFor: htmlFor || controlId, className: classNames(className, bsPrefix) }));
});
FormCheckLabel.displayName = 'FormCheckLabel';
FormCheckLabel.propTypes = propTypes$2;

const propTypes$1 = {
    /**
     * @default 'form-check'
     */
    bsPrefix: PropTypes.string,
    /**
     * bsPrefix override for the base switch class.
     *
     * @default 'form-switch'
     */
    bsSwitchPrefix: PropTypes.string,
    /**
     * The FormCheck `ref` will be forwarded to the underlying input element,
     * which means it will be a DOM node, when resolved.
     *
     * @type {ReactRef}
     * @alias ref
     */
    _ref: PropTypes.any,
    /**
     * The underlying HTML element to use when rendering the FormCheck.
     *
     * @type {('input'|elementType)}
     */
    as: PropTypes.elementType,
    /**
     * A HTML id attribute, necessary for proper form accessibility.
     * An id is recommended for allowing label clicks to toggle the check control.
     *
     * This is **required** when `type="switch"` due to how they are rendered.
     */
    id: PropTypes.string,
    /**
     * Groups controls horizontally with other `FormCheck`s.
     */
    inline: PropTypes.bool,
    /**
     * Disables the control.
     */
    disabled: PropTypes.bool,
    /**
     * `title` attribute for the underlying `FormCheckLabel`.
     */
    title: PropTypes.string,
    /**
     * Label for the control.
     */
    label: PropTypes.node,
    /**
     * aria-label for label element.
     */
    ariaLabel: PropTypes.string,
    /**
     * The type of checkable.
     * @type {('radio' | 'checkbox' | 'switch')}
     */
    type: PropTypes.oneOf(['radio', 'checkbox', 'switch']),
    /** Manually style the input as valid */
    isValid: PropTypes.bool,
    /** Manually style the input as invalid */
    isInvalid: PropTypes.bool,
    /** A message to display when the input is in a validation state */
    feedback: PropTypes.node,
    feedbackType: PropTypes.oneOf(['invalid', 'valid'])
};
const FormCheck = React.forwardRef(({ id, bsPrefix, bsSwitchPrefix, inline = false, disabled = false, isValid = false, isInvalid = false, feedback, feedbackType, className, style, title = '', type = 'checkbox', label, ariaLabel, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as = 'input', ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'form-check');
    bsSwitchPrefix = useBootstrapPrefix(bsSwitchPrefix, 'form-switch');
    const { controlId } = useContext(FormContext);
    const innerFormContext = useMemo(() => ({
        controlId: id || controlId,
    }), [controlId, id]);
    const input = (React.createElement(FormCheckInput, { ...props, type: type === 'switch' ? 'checkbox' : type, ref: ref, isValid: isValid, isInvalid: isInvalid, disabled: disabled, as: as }));
    return (React.createElement(FormContext.Provider, { value: innerFormContext },
        React.createElement("div", { style: style, className: classNames(className, label && bsPrefix, inline && `${bsPrefix}-inline`, type === 'switch' && bsSwitchPrefix) },
            input,
            React.createElement(FormCheckLabel, { title: title, "aria-label": ariaLabel }, label),
            feedback && React.createElement(Feedback, { type: feedbackType }, feedback))));
});
FormCheck.displayName = 'FormCheck';
FormCheck.propTypes = propTypes$1;
var FormCheck$1 = Object.assign(FormCheck, {
    Input: FormCheckInput,
    Label: FormCheckLabel,
});

const propTypes = {
    /**
     * The type of checkable.
     * @type {('radio' | 'checkbox' | 'switch')}
     */
    type: PropTypes.oneOf(['radio', 'checkbox', 'switch']),
    /**
     * Disables the control.
     */
    disabled: PropTypes.bool,
    /**
     * Sets card background
     *
     * @type {('primary'|'secondary'|'success'|'danger'|'warning'|'info'|'dark'|'light')}
     */
    bg: PropTypes.string,
    /**
     * Sets card text color
     *
     * @type {('primary'|'secondary'|'success'|'danger'|'warning'|'info'|'dark'|'light'|'white'|'muted')}
     */
    text: PropTypes.string,
    /**
     * Sets card border color
     *
     * @type {('primary'|'secondary'|'success'|'danger'|'warning'|'info'|'dark'|'light')}
     */
    border: PropTypes.string,
    /** Category title of the Card */
    categoryTitle: PropTypes.node
};
const defaultProps = {
    disabled: false,
    type: 'checkbox',
};
const SelectableCard = ({ children, bg, text, border, categoryTitle, ...props }) => {
    const formCheckRef = React.useRef(null);
    const cardRef = React.useRef(null);
    const formCheckProps = props;
    const cardProps = { bg, text, border };
    const handleSelect = () => {
        formCheckRef?.current?.click();
    };
    return (React.createElement(Card$1, { ref: cardRef, onClick: handleSelect, tabIndex: 0, variant: "card-action", className: props.checked && !props.disabled
            ? 'is-active'
            : undefined, ...cardProps },
        React.createElement(Card$1.Body, null,
            React.createElement(Card$1.Subtitle, { as: "h6", className: "text-muted" },
                React.createElement("div", null, categoryTitle),
                React.createElement("div", { className: "card-input" },
                    React.createElement(FormCheck$1, { ref: formCheckRef, ...formCheckProps, onClick: handleSelect }))),
            children)));
};
SelectableCard.displayName = 'SelectableCard';
SelectableCard.defaultProps = defaultProps;
SelectableCard.propTypes = propTypes;

var CardGroup = createWithBsPrefix('card-group');

export { Card$1 as Card, CardBody, CardFooter, CardGroup, CardHeader, CardImg, CardImgOverlay, CardLink, CardStretchedLink, CardSubtitle, CardText, CardTitle, CardUnit, SelectableCard };
