// @govtechsg/sgds-react/Footer/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var classNames = require('classnames');
var React = require('react');
var PropTypes = require('prop-types');
var warning = require('warning');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__namespace = /*#__PURE__*/_interopNamespace(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);

const ThemeContext = React__namespace.createContext({ prefixes: {} });
({
    prefixes: PropTypes__default["default"].object,
    dir: PropTypes__default["default"].string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = React.useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
const SGDSWrapper = React__namespace.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React__namespace.createElement(Tag, { ref: ref, ...props, className: classNames__default["default"](props.className, 'sgds') });
});

const propTypes$5 = {
    as: PropTypes__default["default"].elementType,
    bsPrefix: PropTypes__default["default"].string,
};
const FooterBottomCopyrights = React__namespace.forwardRef(({ as: Component = 'div', children, className, bsPrefix, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer-copyrights');
    return (React__namespace.createElement(Component, { ref: ref, className: classNames__default["default"](className, bsPrefix, 'row'), ...props },
        React__namespace.createElement("div", { className: "col" },
            React__namespace.createElement("div", { className: "d-flex justify-content-lg-end" }, children))));
});
FooterBottomCopyrights.propTypes = propTypes$5;
FooterBottomCopyrights.displayName = 'FooterBottomCopyrights';

const propTypes$4 = {
    as: PropTypes__default["default"].elementType,
    bsPrefix: PropTypes__default["default"].string
};
const FooterBottomLinks = React__namespace.forwardRef(({ as: Component = 'div', children, className, bsPrefix, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer-mandatory-links');
    return (React__namespace.createElement(Component, { ref: ref, className: classNames__default["default"](className, bsPrefix, 'row'), ...props },
        React__namespace.createElement("div", { className: "col" },
            React__namespace.createElement("ul", null, React__namespace.Children.map(children, (child) => (React__namespace.createElement("li", null, React__namespace.cloneElement(child))))))));
});
FooterBottomLinks.propTypes = propTypes$4;
FooterBottomLinks.displayName = 'FooterBottomLinks';

const FooterBottom = React__namespace.forwardRef(({ as: Component = 'section', children, bsPrefix, className, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer-bottom');
    return (React__namespace.createElement(Component, { ref: ref, className: classNames__default["default"](className, bsPrefix), ...props },
        React__namespace.createElement("div", { className: "container-fluid" }, children)));
});
var FooterBottom$1 = Object.assign(FooterBottom, {
    Links: FooterBottomLinks,
    Copyrights: FooterBottomCopyrights
});

const FooterTopContactLinks = React__namespace.forwardRef(({ as: Component = 'div', children, bsPrefix, className, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer-contact-links');
    return (React__namespace.createElement(Component, { className: classNames__default["default"](bsPrefix, 'row', className), ref: ref, ...props },
        React__namespace.createElement("div", { className: "col" },
            React__namespace.createElement("div", { className: "d-flex justify-content-lg-end" },
                React__namespace.createElement("ul", null,
                    " ",
                    React__namespace.Children.map(children, (child) => (React__namespace.createElement("li", null, React__namespace.cloneElement(child)))))))));
});

const propTypes$3 = {
    headerTitle: PropTypes__default["default"].node,
    titleClass: PropTypes__default["default"].string,
    descriptionClass: PropTypes__default["default"].string,
    columnClass: PropTypes__default["default"].string,
    as: PropTypes__default["default"].elementType,
    bsPrefix: PropTypes__default["default"].string
};
const defaultProps = {
    columnClass: 'col-lg-6',
};
const FooterTopHeader = React.forwardRef(({ as: Component = 'div', headerTitle, children, bsPrefix, className, titleClass, descriptionClass, columnClass = 'col-lg-6', ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer-header');
    return (React__namespace.createElement(Component, { className: classNames__default["default"](bsPrefix, className, 'row'), ref: ref, ...props },
        React__namespace.createElement("div", { className: classNames__default["default"]('col', columnClass) },
            React__namespace.createElement("div", { className: classNames__default["default"]('title', titleClass) }, headerTitle),
            React__namespace.createElement("div", { className: classNames__default["default"]('description', descriptionClass) }, children))));
});
FooterTopHeader.defaultProps = defaultProps;
FooterTopHeader.propTypes = propTypes$3;
FooterTopHeader.displayName = 'FooterTopHeader';

const context = React__namespace.createContext({
    noOfItem: 0
});
context.displayName = 'FooterTopItemGrpContext';

const propTypes$2 = {
    itemTitle: PropTypes__default["default"].node,
    titleClass: PropTypes__default["default"].string,
    linksClass: PropTypes__default["default"].string,
    as: PropTypes__default["default"].elementType,
};
const FooterTopItem = React__namespace.forwardRef(({ as: Component = 'div', itemTitle, children, className, titleClass, linksClass, ...props }, ref) => {
    const { noOfItem } = React__namespace.useContext(context);
    const colSize = noOfItem < 5 ? 'col-lg-3' : 'col-lg-2';
    return (React__namespace.createElement(Component, { className: classNames__default["default"](colSize, className), ref: ref, ...props },
        React__namespace.createElement("div", { className: classNames__default["default"]('title', titleClass) }, itemTitle),
        React__namespace.createElement("ul", { className: classNames__default["default"]('links', linksClass) }, React__namespace.Children.map(children, (child) => (React__namespace.createElement("li", null, React__namespace.cloneElement(child)))))));
});
FooterTopItem.displayName = 'FooterTopItem';
FooterTopItem.propTypes = propTypes$2;

const propTypes$1 = {
    as: PropTypes__default["default"].elementType,
    bsPrefix: PropTypes__default["default"].string,
};
const FooterTopItemGroup = React__namespace.forwardRef(({ as: Component = 'div', children, className, bsPrefix, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer-items');
    const noOfChildren = React__namespace.Children.count(children);
    const contextValue = React__namespace.useMemo(() => ({ noOfItem: noOfChildren }), [noOfChildren]);
    warning__default["default"](noOfChildren <= 6, 'Number of Footer Items should not exceed 6');
    return (React__namespace.createElement(context.Provider, { value: contextValue },
        React__namespace.createElement(Component, { className: classNames__default["default"](bsPrefix, className, 'row'), ref: ref, ...props }, children)));
});
FooterTopItemGroup.displayName = 'FooterTopItemGroup';
FooterTopItemGroup.propTypes = propTypes$1;

const FooterTop = React__namespace.forwardRef(({ as: Component = 'section', children, className, bsPrefix, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer-top');
    return (React__namespace.createElement(Component, { ref: ref, className: classNames__default["default"](bsPrefix, className), ...props },
        React__namespace.createElement("div", { className: "container-fluid" }, children)));
});
var FooterSection = Object.assign(FooterTop, {
    Header: FooterTopHeader,
    ItemGroup: FooterTopItemGroup,
    Item: FooterTopItem,
    ContactLinks: FooterTopContactLinks,
});

const propTypes = {
    as: PropTypes__default["default"].elementType,
    bsPrefix: PropTypes__default["default"].string,
};
const Footer = React__namespace.forwardRef(({ as: Component = 'footer', bsPrefix, className, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'footer');
    return (React__namespace.createElement(SGDSWrapper, { as: Component, ref: ref, className: classNames__default["default"](className, bsPrefix), ...props }));
});
Footer.displayName = 'Footer';
Footer.propTypes = propTypes;
var Footer$1 = Object.assign(Footer, {
    Top: FooterSection,
    Bottom: FooterBottom$1,
});

exports.Footer = Footer$1;
exports.FooterBottom = FooterBottom$1;
exports.FooterBottomCopyrights = FooterBottomCopyrights;
exports.FooterBottomLinks = FooterBottomLinks;
exports.FooterTop = FooterSection;
exports.FooterTopContactLinks = FooterTopContactLinks;
exports.FooterTopHeader = FooterTopHeader;
exports.FooterTopItem = FooterTopItem;
exports.FooterTopItemGroup = FooterTopItemGroup;
