// @govtechsg/sgds-react/Nav/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var classNames = require('classnames');
var PropTypes = require('prop-types');
var React = require('react');
var uncontrollable = require('uncontrollable');
var BaseNav = require('@restart/ui/Nav');
var createWithBsPrefix = require('../utils/createWithBsPrefix');
var Anchor = require('@restart/ui/Anchor');
var NavItem$1 = require('@restart/ui/NavItem');
var SelectableContext = require('@restart/ui/SelectableContext');
var BaseDropdown = require('@restart/ui/Dropdown');
var useEventCallback = require('@restart/hooks/useEventCallback');
var DropdownItem$1 = require('@restart/ui/DropdownItem');
var DropdownMenu$1 = require('@restart/ui/DropdownMenu');
var useIsomorphicEffect = require('@restart/hooks/useIsomorphicEffect');
var useMergedRefs = require('@restart/hooks/useMergedRefs');
var warning = require('warning');
var types = require('../utils/types');
var useWrappedRefWithWarning = require('../utils/useWrappedRefWithWarning');
var DropdownToggle$1 = require('@restart/ui/DropdownToggle');
var DropdownContext$1 = require('@restart/ui/DropdownContext');
var Button$1 = require('@restart/ui/Button');
var css = require('dom-helpers/css');
var Transition = require('react-transition-group/Transition');
var transitionEndListener = require('../utils/transitionEndListener');
var createChainedFunction = require('../utils/createChainedFunction');
var triggerBrowserReflow = require('../utils/triggerBrowserReflow');
var TransitionWrapper = require('../utils/TransitionWrapper');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var React__namespace = /*#__PURE__*/_interopNamespace(React);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var BaseNav__default = /*#__PURE__*/_interopDefaultLegacy(BaseNav);
var createWithBsPrefix__default = /*#__PURE__*/_interopDefaultLegacy(createWithBsPrefix);
var Anchor__default = /*#__PURE__*/_interopDefaultLegacy(Anchor);
var SelectableContext__default = /*#__PURE__*/_interopDefaultLegacy(SelectableContext);
var BaseDropdown__default = /*#__PURE__*/_interopDefaultLegacy(BaseDropdown);
var useEventCallback__default = /*#__PURE__*/_interopDefaultLegacy(useEventCallback);
var useIsomorphicEffect__default = /*#__PURE__*/_interopDefaultLegacy(useIsomorphicEffect);
var useMergedRefs__default = /*#__PURE__*/_interopDefaultLegacy(useMergedRefs);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var useWrappedRefWithWarning__default = /*#__PURE__*/_interopDefaultLegacy(useWrappedRefWithWarning);
var DropdownContext__default = /*#__PURE__*/_interopDefaultLegacy(DropdownContext$1);
var css__default = /*#__PURE__*/_interopDefaultLegacy(css);
var transitionEndListener__default = /*#__PURE__*/_interopDefaultLegacy(transitionEndListener);
var createChainedFunction__default = /*#__PURE__*/_interopDefaultLegacy(createChainedFunction);
var triggerBrowserReflow__default = /*#__PURE__*/_interopDefaultLegacy(triggerBrowserReflow);
var TransitionWrapper__default = /*#__PURE__*/_interopDefaultLegacy(TransitionWrapper);

const ThemeContext = React__namespace.createContext({ prefixes: {} });
({
    prefixes: PropTypes__default["default"].object,
    dir: PropTypes__default["default"].string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = React.useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
// `sgds-${defaultPrefix}`
function useIsRTL() {
    const { dir } = React.useContext(ThemeContext);
    return dir === 'rtl';
}
const SGDSWrapper = React__namespace.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React__namespace.createElement(Tag, { ref: ref, ...props, className: classNames__default["default"](props.className, 'sgds') });
});

const context$1 = React__namespace.createContext(null);
context$1.displayName = 'NavbarContext';

const NavItem = createWithBsPrefix__default["default"]('nav-item', {
    Component: "li",
    displayName: 'NavItem'
});

const propTypes$c = {
    /**
     * @default 'nav-link'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * The active state of the NavItem item.
     */
    active: PropTypes__default["default"].bool,
    /**
     * The disabled state of the NavItem item.
     */
    disabled: PropTypes__default["default"].bool,
    /**
     * The ARIA role for the `NavLink`, In the context of a 'tablist' parent Nav,
     * the role defaults to 'tab'
     * */
    role: PropTypes__default["default"].string,
    /** The HTML href attribute for the `NavLink` */
    href: PropTypes__default["default"].string,
    /**
     * Uniquely identifies the `NavItem` amongst its siblings,
     * used to determine and control the active state of the parent `Nav`
     */
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /** @default 'a' */
    as: PropTypes__default["default"].elementType,
};
const defaultProps$6 = {
    disabled: false,
};
const NavLink = React__namespace.forwardRef(({ bsPrefix, className, as: Component = Anchor__default["default"], active, eventKey, 
// onClick,
...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'nav-link');
    const [navItemProps, meta] = NavItem$1.useNavItem({
        key: SelectableContext.makeEventKey(eventKey, props.href),
        active,
        ...props,
    });
    return (React__namespace.createElement(Component, { ...props, ...navItemProps, ref: ref, className: classNames__default["default"](className, bsPrefix, props.disabled && 'disabled', meta.isActive && 'active'), "aria-current": meta.isActive ? "page" : "false" }));
});
NavLink.displayName = 'NavLink';
NavLink.propTypes = propTypes$c;
NavLink.defaultProps = defaultProps$6;

const DropdownContext = React__namespace.createContext({});
DropdownContext.displayName = 'DropdownContext';

const propTypes$b = {
    /** @default 'dropdown-item' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * Highlight the menu item as active.
     */
    active: PropTypes__default["default"].bool,
    /**
     * Disable the menu item, making it unselectable.
     */
    disabled: PropTypes__default["default"].bool,
    /**
     * Value passed to the `onSelect` handler, useful for identifying the selected menu item.
     */
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * HTML `href` attribute corresponding to `a.href`.
     */
    href: PropTypes__default["default"].string,
    /**
     * Callback fired when the menu item is clicked.
     */
    onClick: PropTypes__default["default"].func,
    as: PropTypes__default["default"].elementType,
};
const DropdownItem = React__namespace.forwardRef(({ bsPrefix, className, eventKey, disabled = false, onClick, active, as: Component = Anchor__default["default"], ...props }, ref) => {
    const prefix = 'dropdown-item';
    const [dropdownItemProps, meta] = DropdownItem$1.useDropdownItem({
        key: eventKey,
        href: props.href,
        disabled,
        onClick,
        active,
    });
    return (React__namespace.createElement("li", null,
        React__namespace.createElement(Component, { ...props, ...dropdownItemProps, ref: ref, className: classNames__default["default"](className, prefix, meta.isActive && 'active', disabled && 'disabled') })));
});
DropdownItem.displayName = 'DropdownItem';
DropdownItem.propTypes = propTypes$b;

const context = React__namespace.createContext(null);
context.displayName = 'InputGroupContext';

const propTypes$a = {
    /**
     * @default 'dropdown-menu'
     */
    bsPrefix: PropTypes__default["default"].string,
    /** Controls the visibility of the Dropdown menu  */
    show: PropTypes__default["default"].bool,
    /** Whether to render the dropdown menu in the DOM before the first time it is shown */
    renderOnMount: PropTypes__default["default"].bool,
    /** Have the dropdown switch to it's opposite placement when necessary to stay on screen. */
    flip: PropTypes__default["default"].bool,
    /**
     * Aligns the dropdown menu to the specified side of the container. You can also align
     * the menu responsively for breakpoints starting at `sm` and up. The alignment
     * direction will affect the specified breakpoint or larger.
     *
     * *Note: Using responsive alignment will disable Popper usage for positioning.*
     *
     * @type {"start"|"end"|{ sm: "start"|"end" }|{ md: "start"|"end" }|{ lg: "start"|"end" }|{ xl: "start"|"end"}|{ xxl: "start"|"end"} }
     */
    align: types.alignPropType,
    onSelect: PropTypes__default["default"].func,
    /**
     * Which event when fired outside the component will cause it to be closed
     *
     * *Note: For custom dropdown components, you will have to pass the
     * `rootCloseEvent` to `<RootCloseWrapper>` in your custom dropdown menu
     * component ([similarly to how it is implemented in `<Dropdown.Menu>`](https://github.com/react-bootstrap/react-bootstrap/blob/v0.31.5/src/DropdownMenu.js#L115-L119)).*
     */
    rootCloseEvent: PropTypes__default["default"].oneOf(['click', 'mousedown']),
    /**
     * Control the rendering of the DropdownMenu. All non-menu props
     * (listed here) are passed through to the `as` Component.
     *
     * If providing a custom, non DOM, component. the `show`, `close` and `align` props
     * are also injected and should be handled appropriately.
     */
    as: PropTypes__default["default"].elementType,
    /**
     * A set of popper options and props passed directly to Popper.
     */
    popperConfig: PropTypes__default["default"].object,
    /**
     * Menu color variant.
     *
     * Omitting this will use the default light color.
     */
    variant: PropTypes__default["default"].string,
    /** @private Provides context when used in NavDropdown component  */
    isNav: PropTypes__default["default"].bool,
};
const defaultProps$5 = {
    flip: true,
};
function getDropdownMenuPlacement(alignEnd, dropDirection, isRTL) {
    const topStart = isRTL ? 'top-end' : 'top-start';
    const topEnd = isRTL ? 'top-start' : 'top-end';
    const bottomStart = isRTL ? 'bottom-end' : 'bottom-start';
    const bottomEnd = isRTL ? 'bottom-start' : 'bottom-end';
    // const leftStart = isRTL ? 'right-start' : 'left-start';
    // const leftEnd = isRTL ? 'right-end' : 'left-end';
    const rightStart = isRTL ? 'left-start' : 'right-start';
    const rightEnd = isRTL ? 'left-end' : 'right-end';
    let placement = alignEnd ? bottomEnd : bottomStart;
    if (dropDirection === 'up')
        placement = alignEnd ? topEnd : topStart;
    else if (dropDirection === 'end')
        placement = alignEnd ? rightEnd : rightStart;
    // else if (dropDirection === 'start')
    //   placement = alignEnd ? leftEnd : leftStart;
    return placement;
}
const DropdownMenu = React__namespace.forwardRef(({ bsPrefix, className, align, rootCloseEvent, flip, show: showProps, renderOnMount, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = 'ul', popperConfig, variant, isNav, ...props }, ref) => {
    let alignEnd = false;
    const isNavbar = React.useContext(context$1);
    const prefix = 'dropdown-menu';
    const { align: contextAlign, drop, isRTL } = React.useContext(DropdownContext);
    align = align || contextAlign;
    const isInputGroup = React.useContext(context);
    const alignClasses = [];
    if (align) {
        if (typeof align === 'object') {
            const keys = Object.keys(align);
            warning__default["default"](keys.length === 1, 'There should only be 1 breakpoint when passing an object to `align`');
            if (keys.length) {
                const brkPoint = keys[0];
                const direction = align[brkPoint];
                // .dropdown-menu-end is required for responsively aligning
                // left in addition to align left classes.
                alignEnd = direction === 'start';
                alignClasses.push(`${prefix}-${brkPoint}-${direction}`);
            }
        }
        else if (align === 'end') {
            alignEnd = true;
        }
    }
    const placement = getDropdownMenuPlacement(alignEnd, drop, isRTL);
    const [menuProps, { hasShown, popper, show, toggle }] = DropdownMenu$1.useDropdownMenu({
        flip,
        rootCloseEvent,
        show: showProps,
        usePopper: !isNavbar && alignClasses.length === 0,
        offset: !isNav ? [0, 10] : undefined,
        popperConfig,
        placement,
    });
    menuProps.ref = useMergedRefs__default["default"](useWrappedRefWithWarning__default["default"](ref, 'DropdownMenu'), menuProps.ref);
    useIsomorphicEffect__default["default"](() => {
        // Popper's initial position for the menu is incorrect when
        // renderOnMount=true. Need to call update() to correct it.
        if (show)
            popper?.update();
    }, [show]);
    if (!hasShown && !renderOnMount && !isInputGroup)
        return null;
    // For custom components provide additional, non-DOM, props;
    if (typeof Component !== 'string') {
        menuProps.show = show;
        menuProps.close = () => toggle?.(false);
        menuProps.align = align;
    }
    let style = props.style;
    if (popper?.placement) {
        // we don't need the default popper style,
        // menus are display: none when not shown.
        style = { ...props.style, ...menuProps.style };
        //@ts-ignore
        props['x-placement'] = popper.placement;
    }
    return (React__namespace.createElement(Component, { ...props, ...menuProps, style: style, ...((alignClasses.length || isNavbar) && {
            'data-bs-popper': 'static',
        }), className: classNames__default["default"](className, prefix, show && 'show', alignEnd && `${prefix}-end`, variant && `${prefix}-${variant}`, ...alignClasses) }));
});
DropdownMenu.displayName = 'DropdownMenu';
DropdownMenu.propTypes = propTypes$a;
DropdownMenu.defaultProps = defaultProps$5;

const propTypes$9 = {
    /**
     * @default 'btn'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * One or more button variant combinations
     *
     * buttons may be one of a variety of visual variants such as:
     *
     * `'primary', 'secondary', 'success', 'danger', 'warning', 'info', 'dark', 'light', 'link'`
     *
     * as well as "outline" versions (prefixed by 'outline-*')
     *
     * `'outline-primary', 'outline-secondary', 'outline-success', 'outline-danger', 'outline-warning', 'outline-info', 'outline-dark', 'outline-light'`
     */
    variant: PropTypes__default["default"].string,
    /**
     * Specifies a large or small button.
     *
     * @type ('sm'|'lg')
     */
    size: PropTypes__default["default"].string,
    /** Manually set the visual state of the button to `:active` */
    active: PropTypes__default["default"].bool,
    /**
     * Disables the Button, preventing mouse events,
     * even if the underlying component is an `<a>` element
     */
    disabled: PropTypes__default["default"].bool,
    /** Providing a `href` will render an `<a>` element, _styled_ as a button. */
    href: PropTypes__default["default"].string,
    /**
     * Defines HTML button type attribute.
     *
     * @default 'button'
     */
    type: PropTypes__default["default"].oneOf(['button', 'reset', 'submit', null]),
    as: PropTypes__default["default"].elementType,
};
const defaultProps$4 = {
    variant: 'primary',
    active: false,
    disabled: false,
};
const Button = React__namespace.forwardRef(({ as, bsPrefix, variant, size, active, className, ...props }, ref) => {
    const prefix = useBootstrapPrefix(bsPrefix, 'btn');
    const [buttonProps, { tagName }] = Button$1.useButtonProps({
        tagName: as,
        ...props,
    });
    const Component = tagName;
    return (React__namespace.createElement(SGDSWrapper, { as: Component, ...props, ...buttonProps, ref: ref, className: classNames__default["default"](className, prefix, active && 'active', variant && `${prefix}-${variant}`, size && `${prefix}-${size}`, props.href && props.disabled && 'disabled') }));
});
Button.displayName = 'Button';
Button.propTypes = propTypes$9;
Button.defaultProps = defaultProps$4;

const propTypes$8 = {
    /**
     * @default 'dropdown-toggle'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * An html id attribute, necessary for assistive technologies, such as screen readers.
     * @type {string|number}
     */
    id: PropTypes__default["default"].string,
    split: PropTypes__default["default"].bool,
    as: PropTypes__default["default"].elementType,
    /**
     * to passthrough to the underlying button or whatever from DropdownButton
     * @private
     */
    childBsPrefix: PropTypes__default["default"].string,
};
const DropdownToggle = React__namespace.forwardRef(({ bsPrefix, split, className, childBsPrefix, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = Button, ...props }, ref) => {
    const prefix = 'dropdown-toggle';
    const dropdownContext = React.useContext(DropdownContext__default["default"]);
    const isInputGroup = React.useContext(context);
    if (childBsPrefix !== undefined) {
        props.bsPrefix = childBsPrefix;
    }
    const [toggleProps] = DropdownToggle$1.useDropdownToggle();
    toggleProps.ref = useMergedRefs__default["default"](toggleProps.ref, useWrappedRefWithWarning__default["default"](ref, 'DropdownToggle'));
    // This intentionally forwards size and variant (if set) to the
    // underlying component, to allow it to render size and style variants.
    return (React__namespace.createElement(Component, { className: classNames__default["default"](className, prefix, split && `${prefix}-split`, !!isInputGroup && dropdownContext?.show && 'show'), variant: "outline-secondary", ...toggleProps, ...props, "aria-haspopup": "menu" }));
});
DropdownToggle.displayName = 'DropdownToggle';
DropdownToggle.propTypes = propTypes$8;

const DropdownHeader = createWithBsPrefix__default["default"]('dropdown-header', {
    defaultProps: { role: 'heading' },
});
const DropdownDivider = createWithBsPrefix__default["default"]('dropdown-divider', {
    Component: 'hr',
    defaultProps: { role: 'separator' },
});
const DropdownItemText = createWithBsPrefix__default["default"]('dropdown-item-text', {
    Component: 'span',
});

const propTypes$7 = {
    /** @default 'dropdown' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * Determines the direction and location of the Menu in relation to it's Toggle.
     *
     * @type {"end"|"up"|"down" }
     */
    drop: PropTypes__default["default"].oneOf(['up', 'end', 'down']),
    as: PropTypes__default["default"].elementType,
    align: types.alignDirection,
    /**
     * Whether or not the Dropdown is visible.
     *
     * @controllable onToggle
     */
    show: PropTypes__default["default"].bool,
    /**
     * Allow Dropdown to flip in case of an overlapping on the reference element. For more information refer to
     * Popper.js's flip [docs](https://popper.js.org/docs/v2/modifiers/flip/).
     *
     */
    flip: PropTypes__default["default"].bool,
    /**
     * A callback fired when the Dropdown wishes to change visibility. Called with the requested
     * `show` value, the DOM event, and the source that fired it: `'click'`,`'keydown'`,`'rootClose'`, or `'select'`.
     *
     * ```js
     * function(
     *   isOpen: boolean,
     *   event: SyntheticEvent,
     *   metadata: {
     *     source: 'select' | 'click' | 'rootClose' | 'keydown'
     *   }
     * ): void
     * ```
     *
     * @controllable show
     */
    onToggle: PropTypes__default["default"].func,
    /**
     * A callback fired when a menu item is selected.
     *
     * ```js
     * (eventKey: any, event: Object) => any
     * ```
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * Controls the focus behavior for when the Dropdown is opened. Set to
     * `true` to always focus the first menu item, `keyboard` to focus only when
     * navigating via the keyboard, or `false` to disable completely
     *
     * The Default behavior is `false` **unless** the Menu has a `role="menu"`
     * where it will default to `keyboard` to match the recommended [ARIA Authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton).
     */
    focusFirstItemOnShow: PropTypes__default["default"].oneOf([false, true, 'keyboard']),
    /** @private */
    navbar: PropTypes__default["default"].bool,
    /**
     * Controls the auto close behaviour of the dropdown when clicking outside of
     * the button or the list.
     */
    autoClose: PropTypes__default["default"].oneOf([true, 'outside', 'inside', false]),
};
const defaultProps$3 = {
    navbar: false,
    align: 'start',
    autoClose: true,
    focusFirstItemOnShow: 'keyboard',
};
const Dropdown = React__namespace.forwardRef((pProps, ref) => {
    const { bsPrefix, drop, show, className, align, onSelect, onToggle, focusFirstItemOnShow, 
    // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
    as: Component = 'div', navbar: _4, autoClose, ...props } = uncontrollable.useUncontrolled(pProps, { show: 'onToggle' });
    const isInputGroup = React.useContext(context);
    const prefix = useBootstrapPrefix(bsPrefix, 'dropdown');
    const isRTL = useIsRTL();
    const isClosingPermitted = (source) => {
        // autoClose=false only permits close on button click
        if (autoClose === false)
            return source === 'click';
        // autoClose=inside doesn't permit close on rootClose
        if (autoClose === 'inside')
            return source !== 'rootClose';
        // autoClose=outside doesn't permit close on select
        if (autoClose === 'outside')
            return source !== 'select';
        return true;
    };
    const handleToggle = useEventCallback__default["default"]((nextShow, meta) => {
        if (meta.originalEvent.currentTarget === document &&
            (meta.source !== 'keydown' ||
                meta.originalEvent.key === 'Escape'))
            meta.source = 'rootClose';
        if (isClosingPermitted(meta.source))
            onToggle?.(nextShow, meta);
    });
    const alignEnd = align === 'end';
    const placement = getDropdownMenuPlacement(alignEnd, drop, isRTL);
    const contextValue = React.useMemo(() => ({
        align,
        drop,
        isRTL,
    }), [align, drop, isRTL]);
    return (React__namespace.createElement(DropdownContext.Provider, { value: contextValue },
        React__namespace.createElement(BaseDropdown__default["default"], { placement: placement, show: show, onSelect: onSelect, onToggle: handleToggle, focusFirstItemOnShow: focusFirstItemOnShow, itemSelector: `.${prefix}-item:not(.disabled):not(:disabled)` }, isInputGroup ? (props.children) : (React__namespace.createElement(SGDSWrapper, { as: Component, ...props, ref: ref, className: classNames__default["default"](className, show && 'show', prefix, drop === 'up' && 'dropup', drop === 'end' && 'dropend') })))));
});
Dropdown.displayName = 'Dropdown';
Dropdown.propTypes = propTypes$7;
Dropdown.defaultProps = defaultProps$3;
var Dropdown$1 = Object.assign(Dropdown, {
    Toggle: DropdownToggle,
    Menu: DropdownMenu,
    Item: DropdownItem,
    ItemText: DropdownItemText,
    Divider: DropdownDivider,
    Header: DropdownHeader,
});

const NavContext = React__namespace.createContext(null);
NavContext.displayName = 'NavContext';

const propTypes$6 = {
    /**
     * An html id attribute for the Toggle button, necessary for assistive technologies, such as screen readers.
     * @type {string}
     */
    id: PropTypes__default["default"].string,
    /** An `onClick` handler passed to the Toggle component */
    onClick: PropTypes__default["default"].func,
    /** The content of the non-toggle Button.  */
    title: PropTypes__default["default"].node.isRequired,
    /** Disables the toggle NavLink  */
    disabled: PropTypes__default["default"].bool,
    /** Style the toggle NavLink as active  */
    active: PropTypes__default["default"].bool,
    /** An ARIA accessible role applied to the Menu component. When set to 'menu', The dropdown */
    menuRole: PropTypes__default["default"].string,
    /** Whether to render the dropdown menu in the DOM before the first time it is shown */
    renderMenuOnMount: PropTypes__default["default"].bool,
    /**
     *  Which event when fired outside the component will cause it to be closed.
     *
     * _see [DropdownMenu](#menu-props) for more details_
     */
    rootCloseEvent: PropTypes__default["default"].string,
    /**
     * Menu color variant.
     *
     * Omitting this will use the default light color.
     */
    menuVariant: PropTypes__default["default"].oneOf(['dark']),
    /** @ignore */
    bsPrefix: PropTypes__default["default"].string,
    /** When true, applies mega menu stylings */
    isMegaMenu: PropTypes__default["default"].bool,
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
};
const NavDropdown = React__namespace.forwardRef(({ id, title, children, bsPrefix, className, rootCloseEvent, menuRole, disabled, active, renderMenuOnMount, menuVariant, isMegaMenu, eventKey, align, ...props }, ref) => {
    /* NavItem has no additional logic, it's purely presentational. Can set nav item class here to support "as" */
    const navContext = React.useContext(NavContext);
    const navItemPrefix = useBootstrapPrefix(undefined, 'nav-item');
    const dropDownClass = classNames__default["default"](className, navItemPrefix, isMegaMenu ? 'has-megamenu' : undefined);
    return (React__namespace.createElement(Dropdown$1, { ref: ref, as: "li", ...props, className: dropDownClass, align: align },
        React__namespace.createElement(Dropdown$1.Toggle, { id: id, eventKey: null, active: active || navContext?.activeKey === eventKey, disabled: disabled, childBsPrefix: bsPrefix, as: NavLink },
            title,
            React__namespace.createElement("i", { className: "bi bi-chevron-down" })),
        React__namespace.createElement(Dropdown$1.Menu, { role: menuRole, renderOnMount: renderMenuOnMount, rootCloseEvent: rootCloseEvent, variant: menuVariant, isNav: true, align: align, as: isMegaMenu ? 'div' : undefined, className: isMegaMenu ? 'mega-menu' : undefined }, children)));
});
NavDropdown.displayName = 'NavDropdown';
NavDropdown.propTypes = propTypes$6;
var NavDropdown$1 = Object.assign(NavDropdown, {
    Item: Dropdown$1.Item,
    ItemText: Dropdown$1.ItemText,
    Divider: Dropdown$1.Divider,
    Header: Dropdown$1.Header,
});

const propTypes$5 = {
    /**
     * @default 'nav'
     */
    bsPrefix: PropTypes__default["default"].string,
    /** @private */
    // navbarBsPrefix: PropTypes.string,
    /** @private */
    // cardHeaderBsPrefix: PropTypes.string,
    /**
     * The visual variant of the nav items.
     *
     * @type {('tabs-basic-toggle' | 'tabs-info-toggle')}
     */
    variant: PropTypes__default["default"].string,
    /**
     * The default active key that is selected on start.
     */
    defaultActiveKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * Marks the NavItem with a matching `eventKey` (or `href` if present) as active.
     */
    activeKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * A callback fired when a NavItem is selected.
     *
     * ```js
     * function (
     *  Any eventKey,
     *  SyntheticEvent event?
     * )
     * ```
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * ARIA role for the Nav, in the context of a TabContainer, the default will
     * be set to "tablist", but can be overridden by the Nav when set explicitly.
     *
     * When the role is "tablist", NavLink focus is managed according to
     * the ARIA authoring practices for tabs:
     * https://www.w3.org/TR/2013/WD-wai-aria-practices-20130307/#tabpanel
     */
    role: PropTypes__default["default"].string,
    /**
     * Apply styling an alignment for use in a Navbar. This prop will be set
     * automatically when the Nav is used inside a Navbar.
     */
    // navbar: PropTypes.bool,
    /**
     * Enable vertical scrolling within the toggleable contents of a collapsed Navbar.
     */
    navbarScroll: PropTypes__default["default"].bool,
    as: PropTypes__default["default"].elementType,
    /** @private */
    onKeyDown: PropTypes__default["default"].func,
};
const Nav = React__namespace.forwardRef((uncontrolledProps, ref) => {
    const { as = 'ul', bsPrefix: initialBsPrefix, navbarScroll, className, activeKey, ...props } = uncontrollable.useUncontrolled(uncontrolledProps, { activeKey: 'onSelect' });
    const bsPrefix = useBootstrapPrefix(initialBsPrefix, 'nav');
    const contextValue = React.useMemo(() => ({
        activeKey,
    }), [activeKey]);
    const navbarContext = React.useContext(context$1);
    return (React__namespace.createElement(NavContext.Provider, { value: contextValue },
        React__namespace.createElement(BaseNav__default["default"], { as: as, ref: ref, activeKey: activeKey, className: classNames__default["default"](className, {
                [bsPrefix]: !navbarContext,
                [`${navbarContext?.bsPrefix}-nav`]: navbarContext,
                [`${navbarContext?.bsPrefix}-nav-scroll`]: navbarContext && navbarScroll,
                ['sgds']: !navbarContext,
                ['nav-tabs']: !navbarContext,
            }), ...props })));
});
Nav.displayName = 'Nav';
Nav.propTypes = propTypes$5;
var Nav$1 = Object.assign(Nav, {
    Item: NavItem,
    Link: NavLink,
    Dropdown: NavDropdown$1,
});

const propTypes$4 = {
    /** @default 'navbar' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * An href, when provided the Brand will render as an `<a>` element (unless `as` is provided).
     */
    href: PropTypes__default["default"].string,
    /**
     * Set a custom element for this component.
     */
    as: PropTypes__default["default"].elementType,
};
const NavbarBrand = React__namespace.forwardRef(({ bsPrefix, className, as, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'navbar-brand');
    const Component = as || (props.href ? 'a' : 'span');
    return (React__namespace.createElement(Component, { ...props, ref: ref, className: classNames__default["default"](className, bsPrefix) }));
});
NavbarBrand.displayName = 'NavbarBrand';
NavbarBrand.propTypes = propTypes$4;

const MARGINS = {
    height: ['marginTop', 'marginBottom'],
    width: ['marginLeft', 'marginRight'],
};
function getDefaultDimensionValue(dimension, elem) {
    const offset = `offset${dimension[0].toUpperCase()}${dimension.slice(1)}`;
    const value = elem[offset];
    const margins = MARGINS[dimension];
    return (value +
        // @ts-ignore
        parseInt(css__default["default"](elem, margins[0]), 10) +
        // @ts-ignore
        parseInt(css__default["default"](elem, margins[1]), 10));
}
const collapseStyles = {
    [Transition.EXITED]: 'collapse',
    [Transition.EXITING]: 'collapsing',
    [Transition.ENTERING]: 'collapsing',
    [Transition.ENTERED]: 'collapse show',
};
const propTypes$3 = {
    /**
     * Show the component; triggers the expand or collapse animation
     */
    in: PropTypes__default["default"].bool,
    /**
     * Wait until the first "enter" transition to mount the component (add it to the DOM)
     */
    mountOnEnter: PropTypes__default["default"].bool,
    /**
     * Unmount the component (remove it from the DOM) when it is collapsed
     */
    unmountOnExit: PropTypes__default["default"].bool,
    /**
     * Run the expand animation when the component mounts, if it is initially
     * shown
     */
    appear: PropTypes__default["default"].bool,
    /**
     * Duration of the collapse animation in milliseconds, to ensure that
     * finishing callbacks are fired even if the original browser transition end
     * events are canceled
     */
    timeout: PropTypes__default["default"].number,
    /**
     * Callback fired before the component expands
     */
    onEnter: PropTypes__default["default"].func,
    /**
     * Callback fired after the component starts to expand
     */
    onEntering: PropTypes__default["default"].func,
    /**
     * Callback fired after the component has expanded
     */
    onEntered: PropTypes__default["default"].func,
    /**
     * Callback fired before the component collapses
     */
    onExit: PropTypes__default["default"].func,
    /**
     * Callback fired after the component starts to collapse
     */
    onExiting: PropTypes__default["default"].func,
    /**
     * Callback fired after the component has collapsed
     */
    onExited: PropTypes__default["default"].func,
    /**
     * The dimension used when collapsing, or a function that returns the
     * dimension
     */
    dimension: PropTypes__default["default"].oneOfType([
        PropTypes__default["default"].oneOf(['height', 'width']),
        PropTypes__default["default"].func,
    ]),
    /**
     * Function that returns the height or width of the animating DOM node
     *
     * Allows for providing some custom logic for how much the Collapse component
     * should animate in its specified dimension. Called with the current
     * dimension prop value and the DOM node.
     *
     * @default element.offsetWidth | element.offsetHeight
     */
    getDimensionValue: PropTypes__default["default"].func,
    /**
     * ARIA role of collapsible element
     */
    role: PropTypes__default["default"].string,
    /**
     * You must provide a single JSX child element to this component and that element cannot be a \<React.Fragment\>
     */
    children: PropTypes__default["default"].element.isRequired,
};
const defaultProps$2 = {
    in: false,
    timeout: 300,
    mountOnEnter: false,
    unmountOnExit: false,
    appear: false,
    getDimensionValue: getDefaultDimensionValue,
};
const Collapse = React__default["default"].forwardRef(({ onEnter, onEntering, onEntered, onExit, onExiting, className, children, dimension = 'height', getDimensionValue = getDefaultDimensionValue, ...props }, ref) => {
    /* Compute dimension */
    const computedDimension = typeof dimension === 'function' ? dimension() : dimension;
    /* -- Expanding -- */
    const handleEnter = React.useMemo(() => createChainedFunction__default["default"]((elem) => {
        elem.style[computedDimension] = '0';
    }, onEnter), [computedDimension, onEnter]);
    const handleEntering = React.useMemo(() => createChainedFunction__default["default"]((elem) => {
        const scroll = `scroll${computedDimension[0].toUpperCase()}${computedDimension.slice(1)}`;
        elem.style[computedDimension] = `${elem[scroll]}px`;
    }, onEntering), [computedDimension, onEntering]);
    const handleEntered = React.useMemo(() => createChainedFunction__default["default"]((elem) => {
        elem.style[computedDimension] = null;
    }, onEntered), [computedDimension, onEntered]);
    /* -- Collapsing -- */
    const handleExit = React.useMemo(() => createChainedFunction__default["default"]((elem) => {
        elem.style[computedDimension] = `${getDimensionValue(computedDimension, elem)}px`;
        triggerBrowserReflow__default["default"](elem);
    }, onExit), [onExit, getDimensionValue, computedDimension]);
    const handleExiting = React.useMemo(() => createChainedFunction__default["default"]((elem) => {
        elem.style[computedDimension] = null;
    }, onExiting), [computedDimension, onExiting]);
    return (React__default["default"].createElement(TransitionWrapper__default["default"], { ref: ref, addEndListener: transitionEndListener__default["default"], ...props, "aria-expanded": props.role ? props.in : null, onEnter: handleEnter, onEntering: handleEntering, onEntered: handleEntered, onExit: handleExit, onExiting: handleExiting, childRef: children.ref }, (state, innerProps) => React__default["default"].cloneElement(children, {
        ...innerProps,
        className: classNames__default["default"](className, children.props.className, collapseStyles[state], computedDimension === 'width' && 'collapse-horizontal'),
    })));
});
// @ts-ignore
Collapse.propTypes = propTypes$3;
// @ts-ignore
Collapse.defaultProps = defaultProps$2;

const propTypes$2 = {
    /** @default 'navbar-collapse' */
    bsPrefix: PropTypes__default["default"].string,
};
const NavbarCollapse = React__namespace.forwardRef(({ children, bsPrefix, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'navbar-collapse');
    const context = React.useContext(context$1);
    return (React__namespace.createElement(Collapse, { in: !!(context && context.expanded), ...props },
        React__namespace.createElement("div", { ref: ref, className: bsPrefix }, children)));
});
NavbarCollapse.displayName = 'NavbarCollapse';
NavbarCollapse.propTypes = propTypes$2;

const propTypes$1 = {
    /** @default 'navbar-toggler' */
    bsPrefix: PropTypes__default["default"].string,
    /** An accessible ARIA label for the toggler button. */
    label: PropTypes__default["default"].string,
    /** @private */
    onClick: PropTypes__default["default"].func,
    /**
     * The toggle content. When empty, the default toggle will be rendered.
     */
    children: PropTypes__default["default"].node,
    as: PropTypes__default["default"].elementType,
};
const defaultProps$1 = {
    label: 'Toggle navigation',
};
const NavbarToggle = React__namespace.forwardRef(({ bsPrefix, className, children, label, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = 'button', onClick, ...props }, ref) => {
    bsPrefix = useBootstrapPrefix(bsPrefix, 'navbar-toggler');
    const { onToggle, expanded } = React.useContext(context$1) || {};
    const handleClick = useEventCallback__default["default"]((e) => {
        if (onClick)
            onClick(e);
        if (onToggle)
            onToggle();
    });
    if (Component === 'button') {
        props.type = 'button';
    }
    return (React__namespace.createElement(Component, { ...props, ref: ref, onClick: handleClick, "aria-label": label, className: classNames__default["default"](className, bsPrefix, !expanded && 'collapsed'), "aria-expanded": expanded }, children || React__namespace.createElement("span", { className: `${bsPrefix}-icon` })));
});
NavbarToggle.displayName = 'NavbarToggle';
NavbarToggle.propTypes = propTypes$1;
NavbarToggle.defaultProps = defaultProps$1;

const NavbarText = createWithBsPrefix__default["default"]('navbar-text', {
    Component: 'span',
});

const propTypes = {
    /** @default 'navbar' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * The breakpoint, below which, the Navbar will collapse.
     * When `true` the Navbar will always be expanded regardless of screen size.
     */
    expand: PropTypes__default["default"].oneOfType([
        PropTypes__default["default"].bool,
        PropTypes__default["default"].oneOf(['sm', 'md', 'lg', 'xl', 'xxl']),
    ]).isRequired,
    /**
     * A convenience prop for adding `bg-*` utility classes since they are so commonly used here.
     * `light` and `dark` are common choices but any `bg-*` class is supported, including any custom ones you might define.
     *
     * Pairs nicely with the `variant` prop.
     */
    bg: PropTypes__default["default"].string,
    /**
     * Create a fixed navbar along the top or bottom of the screen, that scrolls with the
     * page. A convenience prop for the `fixed-*` positioning classes.
     */
    fixed: PropTypes__default["default"].oneOf(['top', 'bottom']),
    /**
     * Position the navbar at the top of the viewport, but only after scrolling past it.
     * A convenience prop for the `sticky-top` positioning class.
     *
     *  __Not supported in <= IE11 and other older browsers without a polyfill__
     */
    sticky: PropTypes__default["default"].oneOf(['top']),
    /**
     * Set a custom element for this component.
     */
    as: PropTypes__default["default"].elementType,
    /**
     * A callback fired when the `<Navbar>` body collapses or expands. Fired when
     * a `<Navbar.Toggle>` is clicked and called with the new `expanded`
     * boolean value.
     *
     * @controllable expanded
     */
    onToggle: PropTypes__default["default"].func,
    /**
     * A callback fired when a descendant of a child `<Nav>` is selected. Should
     * be used to execute complex closing or other miscellaneous actions desired
     * after selecting a descendant of `<Nav>`. Does nothing if no `<Nav>` or `<Nav>`
     * descendants exist. The callback is called with an eventKey, which is a
     * prop from the selected `<Nav>` descendant, and an event.
     *
     * ```js
     * function (
     *  eventKey: mixed,
     *  event?: SyntheticEvent
     * )
     * ```
     *
     * For basic closing behavior after all `<Nav>` descendant onSelect events in
     * mobile viewports, try using collapseOnSelect.
     *
     * Note: If you are manually closing the navbar using this `OnSelect` prop,
     * ensure that you are setting `expanded` to false and not *toggling* between
     * true and false.
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * Toggles `expanded` to `false` after the onSelect event of a descendant of a
     * child `<Nav>` fires. Does nothing if no `<Nav>` or `<Nav>` descendants exist.
     *
     * Manually controlling `expanded` via the onSelect callback is recommended instead,
     * for more complex operations that need to be executed after
     * the `select` event of `<Nav>` descendants.
     */
    collapseOnSelect: PropTypes__default["default"].bool,
    /**
     * Controls the visiblity of the navbar body
     *
     * @controllable onToggle
     */
    expanded: PropTypes__default["default"].bool,
    /**
     * The ARIA role for the navbar, will default to 'navigation' for
     * Navbars whose `as` is something other than `<nav>`.
     *
     * @default 'navigation'
     */
    role: PropTypes__default["default"].string,
    hasBorderBottom: PropTypes__default["default"].bool,
};
const defaultProps = {
    expand: true,
    collapseOnSelect: false,
};
const Navbar = React__namespace.forwardRef((props, ref) => {
    const { bsPrefix: initialBsPrefix, expand, bg, fixed, sticky, className, hasBorderBottom, 
    // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
    as: Component = 'nav', expanded, onToggle, onSelect, collapseOnSelect, ...controlledProps } = uncontrollable.useUncontrolled(props, {
        expanded: 'onToggle',
    });
    const bsPrefix = useBootstrapPrefix(initialBsPrefix, 'navbar');
    const handleCollapse = React.useCallback((...args) => {
        onSelect?.(...args);
        if (collapseOnSelect && expanded) {
            onToggle?.(false);
        }
    }, [onSelect, collapseOnSelect, expanded, onToggle]);
    // will result in some false positives but that seems better
    // than false negatives. strict `undefined` check allows explicit
    // "nulling" of the role if the user really doesn't want one
    if (controlledProps.role === undefined && Component !== 'nav') {
        controlledProps.role = 'navigation';
    }
    let expandClass = `${bsPrefix}-expand`;
    if (typeof expand === 'string')
        expandClass = `${expandClass}-${expand}`;
    const navbarContext = React.useMemo(() => ({
        onToggle: () => onToggle?.(!expanded),
        bsPrefix,
        expanded: !!expanded,
    }), [bsPrefix, expanded, onToggle]);
    return (React__namespace.createElement(context$1.Provider, { value: navbarContext },
        React__namespace.createElement(SelectableContext__default["default"].Provider, { value: handleCollapse },
            React__namespace.createElement(SGDSWrapper, { as: Component, ref: ref, ...controlledProps, className: classNames__default["default"](className, bsPrefix, expand && expandClass, 'navbar-light', bg && `bg-${bg}`, sticky && `sticky-${sticky}`, fixed && `fixed-${fixed}`, hasBorderBottom && 'border-bottom') }))));
});
Navbar.propTypes = propTypes;
Navbar.defaultProps = defaultProps;
Navbar.displayName = 'Navbar';
var Navbar$1 = Object.assign(Navbar, {
    Brand: NavbarBrand,
    Collapse: NavbarCollapse,
    // Offcanvas: NavbarOffcanvas,
    Text: NavbarText,
    Toggle: NavbarToggle,
});

exports.Nav = Nav$1;
exports.NavDropdown = NavDropdown$1;
exports.NavItem = NavItem;
exports.NavLink = NavLink;
exports.Navbar = Navbar$1;
exports.NavbarBrand = NavbarBrand;
exports.NavbarCollapse = NavbarCollapse;
exports.NavbarText = NavbarText;
exports.NavbarToggle = NavbarToggle;
