// @govtechsg/sgds-react/ProgressBar/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var classNames = require('classnames');
var React = require('react');
var PropTypes = require('prop-types');
var ElementChildren = require('../utils/ElementChildren');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__namespace = /*#__PURE__*/_interopNamespace(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

const ThemeContext = React__namespace.createContext({ prefixes: {} });
({
    prefixes: PropTypes__default["default"].object,
    dir: PropTypes__default["default"].string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = React.useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
const SGDSWrapper = React__namespace.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React__namespace.createElement(Tag, { ref: ref, ...props, className: classNames__default["default"](props.className, 'sgds') });
});

const ROUND_PRECISION = 1000;
/**
 * Validate that children, if any, are instances of `<ProgressBar>`.
 */
function onlyProgressBar(props, propName, componentName) {
    const children = props[propName];
    if (!children) {
        return null;
    }
    let error = null;
    React__namespace.Children.forEach(children, (child) => {
        if (error) {
            return;
        }
        /**
         * Compare types in a way that works with libraries that patch and proxy
         * components like react-hot-loader.
         *
         * see https://github.com/gaearon/react-hot-loader#checking-element-types
         */
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        const element = React__namespace.createElement(ProgressBar, null);
        if (child.type === element.type)
            return;
        const childType = child.type;
        const childIdentifier = React__namespace.isValidElement(child)
            ? childType.displayName || childType.name || childType
            : child;
        error = new Error(`Children of ${componentName} can contain only ProgressBar ` +
            `components. Found ${childIdentifier}.`);
    });
    return error;
}
const propTypes = {
    /**
     * Minimum value progress can begin from
     */
    min: PropTypes__default["default"].number,
    /**
     * Current value of progress
     */
    now: PropTypes__default["default"].number,
    /**
     * Maximum value progress can reach
     */
    max: PropTypes__default["default"].number,
    /**
     * Show label that represents visual percentage.
     * EG. 60%
     */
    label: PropTypes__default["default"].node,
    /**
     * Forwards to aria-label attribute on element with role="progressbar"
     *
     */
    ariaLabel: PropTypes__default["default"].string,
    /**
     * Hide's the label visually.
     */
    visuallyHidden: PropTypes__default["default"].bool,
    /**
     * Uses a gradient to create a striped effect.
     */
    striped: PropTypes__default["default"].bool,
    /**
     * Animate's the stripes from right to left
     */
    animated: PropTypes__default["default"].bool,
    /**
     * @private
     * @default 'progress-bar'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * Sets the background class of the progress bar.
     *
     * @type ('success'|'danger'|'warning'|'info')
     */
    variant: PropTypes__default["default"].string,
    /**
     * Child elements (only allows elements of type <ProgressBar />)
     */
    children: onlyProgressBar,
    /**
     * @private
     */
    isChild: PropTypes__default["default"].bool,
};
const defaultProps = {
    min: 0,
    max: 100,
    animated: false,
    isChild: false,
    visuallyHidden: false,
    striped: false,
    ariaLabel: 'progressbar',
};
function getPercentage(now, min, max) {
    const percentage = ((now - min) / (max - min)) * 100;
    return Math.round(percentage * ROUND_PRECISION) / ROUND_PRECISION;
}
function renderProgressBar({ min, now, max, label, visuallyHidden, striped, animated, className, style, variant, bsPrefix, ariaLabel = 'progressbar', ...props }, ref) {
    return (React__namespace.createElement("div", { ref: ref, ...props, role: "progressbar", className: classNames__default["default"](className, `${bsPrefix}-bar`, {
            [`bg-${variant}`]: variant,
            [`${bsPrefix}-bar-animated`]: animated,
            [`${bsPrefix}-bar-striped`]: animated || striped,
        }), style: { width: `${getPercentage(now, min, max)}%`, ...style }, "aria-valuenow": now, "aria-valuemin": min, "aria-valuemax": max, "aria-label": ariaLabel }, visuallyHidden ? (React__namespace.createElement("span", { className: "visually-hidden" }, label)) : (label)));
}
renderProgressBar.propTypes = propTypes;
const ProgressBar = React__namespace.forwardRef(({ isChild, ...props }, ref) => {
    props.bsPrefix = useBootstrapPrefix(props.bsPrefix, 'progress');
    if (isChild) {
        return renderProgressBar(props, ref);
    }
    const { min, now, max, label, visuallyHidden, striped, animated, bsPrefix, variant, className, children, ariaLabel, ...wrapperProps } = props;
    return (React__namespace.createElement(SGDSWrapper, { ref: ref, ...wrapperProps, className: classNames__default["default"](className, bsPrefix) }, children
        ? ElementChildren.map(children, (child) => React.cloneElement(child, { isChild: true }))
        : renderProgressBar({
            min,
            now,
            max,
            label,
            visuallyHidden,
            striped,
            animated,
            bsPrefix,
            variant,
            ariaLabel,
        }, ref)));
});
ProgressBar.displayName = 'ProgressBar';
ProgressBar.propTypes = propTypes;
ProgressBar.defaultProps = defaultProps;

exports.ProgressBar = ProgressBar;
