// @govtechsg/sgds-react/Alert/index.esm.js
'use client';
import classNames from 'classnames';
import * as React from 'react';
import { useContext, useCallback } from 'react';
import PropTypes from 'prop-types';
import { elementType } from 'prop-types-extra';
import { useUncontrolled } from 'uncontrollable';
import useEventCallback from '@restart/hooks/useEventCallback';
import { ENTERING, ENTERED } from 'react-transition-group/Transition';
import transitionEndListener from '../utils/transitionEndListener';
import triggerBrowserReflow from '../utils/triggerBrowserReflow';
import TransitionWrapper from '../utils/TransitionWrapper';
import createWithBsPrefix from '../utils/createWithBsPrefix';
import divWithClassName from '../utils/divWithClassName';

const ThemeContext = React.createContext({ prefixes: {} });
({
    prefixes: PropTypes.object,
    dir: PropTypes.string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
const SGDSWrapper = React.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React.createElement(Tag, { ref: ref, ...props, className: classNames(props.className, 'sgds') });
});

const propTypes$2 = {
    /**
     * Show the component; triggers the fade in or fade out animation
     */
    in: PropTypes.bool,
    /**
     * Wait until the first "enter" transition to mount the component (add it to the DOM)
     */
    mountOnEnter: PropTypes.bool,
    /**
     * Unmount the component (remove it from the DOM) when it is faded out
     */
    unmountOnExit: PropTypes.bool,
    /**
     * Run the fade in animation when the component mounts, if it is initially
     * shown
     */
    appear: PropTypes.bool,
    /**
     * Duration of the fade animation in milliseconds, to ensure that finishing
     * callbacks are fired even if the original browser transition end events are
     * canceled
     */
    timeout: PropTypes.number,
    /**
     * Callback fired before the component fades in
     */
    onEnter: PropTypes.func,
    /**
     * Callback fired after the component starts to fade in
     */
    onEntering: PropTypes.func,
    /**
     * Callback fired after the has component faded in
     */
    onEntered: PropTypes.func,
    /**
     * Callback fired before the component fades out
     */
    onExit: PropTypes.func,
    /**
     * Callback fired after the component starts to fade out
     */
    onExiting: PropTypes.func,
    /**
     * Callback fired after the component has faded out
     */
    onExited: PropTypes.func,
    /**
     * You must provide a single JSX child element to this component and that element cannot be a \<React.Fragment\>
     */
    children: PropTypes.element.isRequired,
    /**
     * Applies additional specified classes during the transition. Takes an object
     * where the keys correspond to the Transition status
     */
    transitionClasses: PropTypes.object,
};
const defaultProps$2 = {
    in: false,
    timeout: 300,
    mountOnEnter: false,
    unmountOnExit: false,
    appear: false,
};
const fadeStyles = {
    [ENTERING]: 'show',
    [ENTERED]: 'show',
};
const Fade = React.forwardRef(({ className, children, transitionClasses = {}, ...props }, ref) => {
    const handleEnter = useCallback((node, isAppearing) => {
        triggerBrowserReflow(node);
        props.onEnter?.(node, isAppearing);
    }, [props]);
    return (React.createElement(TransitionWrapper, { ref: ref, addEndListener: transitionEndListener, ...props, onEnter: handleEnter, childRef: children.ref }, (status, innerProps) => React.cloneElement(children, {
        ...innerProps,
        className: classNames('fade', className, children.props.className, 
        //@ts-ignore
        fadeStyles[status], transitionClasses[status]),
    })));
});
Fade.propTypes = propTypes$2;
Fade.defaultProps = defaultProps$2;
Fade.displayName = 'Fade';

const propTypes$1 = {
    'aria-label': PropTypes.string,
    onClick: PropTypes.func,
    /**
     * Render different color variant for the button.
     *
     * Omitting this will render the default dark color.
     */
    variant: PropTypes.oneOf(['white']),
};
const defaultProps$1 = {
    'aria-label': 'Close',
};
const CloseButton = React.forwardRef(({ className, variant, ...props }, ref) => (React.createElement("button", { ref: ref, type: "button", className: classNames('btn-close', variant && `btn-close-${variant}`, className), ...props })));
CloseButton.displayName = 'CloseButton';
CloseButton.propTypes = propTypes$1;
CloseButton.defaultProps = defaultProps$1;

const AlertLink = createWithBsPrefix('alert-link', {
    Component: 'a',
});

const DivStyledAsH4 = divWithClassName('h4');
const AlertHeading = createWithBsPrefix('alert-heading', {
    Component: DivStyledAsH4,
});

const propTypes = {
    /**
     * @default 'alert'
     */
    bsPrefix: PropTypes.string,
    /**
     * The Alert visual variant
     *
     * @type {'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'light'}
     */
    variant: PropTypes.string,
    /**
     * Renders a properly aligned dismiss button, as well as
     * adding extra horizontal padding to the Alert.
     */
    dismissible: PropTypes.bool,
    /**
     * Controls the visual state of the Alert.
     *
     * @controllable onClose
     */
    show: PropTypes.bool,
    /**
     * Callback fired when alert is closed.
     *
     * @controllable show
     */
    onClose: PropTypes.func,
    /**
     * Sets the text for alert close button.
     */
    closeLabel: PropTypes.string,
    /**
     * Animate the alert dismissal. Defaults to using `<Fade>` animation or use
     * `false` to disable. A custom `react-transition-group` Transition can also
     * be provided.
     */
    transition: PropTypes.oneOfType([PropTypes.bool, elementType]),
};
const defaultProps = {
    variant: 'primary',
    show: true,
    transition: Fade,
    closeLabel: 'Close alert',
};
const Alert = React.forwardRef((uncontrolledProps, ref) => {
    const { bsPrefix, show, as: Component = 'div', closeLabel, className, children, variant, onClose, dismissible, transition, ...props } = useUncontrolled(uncontrolledProps, {
        show: 'onClose',
    });
    const prefix = useBootstrapPrefix(bsPrefix, 'alert');
    const handleClose = useEventCallback((e) => {
        if (onClose) {
            onClose(false, e);
        }
    });
    const Transition = transition === true ? Fade : transition;
    const alert = (React.createElement(SGDSWrapper, { as: Component, role: "alert", ...(!Transition ? props : undefined), ref: ref, className: classNames(className, prefix, variant && `${prefix}-${variant}`, dismissible && `${prefix}-dismissible`) },
        children,
        dismissible && (React.createElement(CloseButton, { onClick: handleClose, "aria-label": closeLabel, className: `btn-sm` }))));
    if (!Transition)
        return show ? alert : null;
    return (React.createElement(Transition, { unmountOnExit: true, ...props, ref: undefined, in: show }, alert));
});
Alert.displayName = 'Alert';
Alert.defaultProps = defaultProps;
Alert.propTypes = propTypes;
var Alert$1 = Object.assign(Alert, {
    Link: AlertLink,
    Heading: AlertHeading,
});

export { Alert$1 as Alert, AlertHeading, AlertLink };
