// @govtechsg/sgds-react/Table/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var classNames = require('classnames');
var React = require('react');
var PropTypes = require('prop-types');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__namespace = /*#__PURE__*/_interopNamespace(React);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);

const ThemeContext = React__namespace.createContext({ prefixes: {} });
({
    prefixes: PropTypes__default["default"].object,
    dir: PropTypes__default["default"].string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = React.useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
const SGDSWrapper = React__namespace.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React__namespace.createElement(Tag, { ref: ref, ...props, className: classNames__default["default"](props.className, 'sgds') });
});

const TableRow = React__namespace.forwardRef(({ as: Component = 'tr', ...props }, ref) => {
    return (React__namespace.createElement(Component, { ...props, ref: ref }));
});
TableRow.displayName = 'TableRow';

const TableHeader = React__namespace.forwardRef(({ as: Component = 'thead', ...props }, ref) => {
    return (React__namespace.createElement(Component, { ...props, ref: ref }));
});
TableHeader.displayName = 'TableHeader';

const TableHeaderCell = React__namespace.forwardRef(({ as: Component = 'th', ...props }, ref) => {
    return (React__namespace.createElement(Component, { ...props, ref: ref }));
});
TableHeaderCell.displayName = 'TableHeaderCell';

const propTypes$1 = {
    /**
     * Handles the click event on the sort label. User is expected to pass in the sort
     * handler function that will sort the respective table column.
     *
     * @param e The click event.
     */
    onClick: PropTypes__default["default"].func,
    /**
     * Whether sorting is currently active on the column.
     */
    active: PropTypes__default["default"].bool,
    /**
     * Whether the column is currently sorted by ascending ('asc') or descending ('desc') order.
     */
    direction: PropTypes__default["default"].oneOf(['asc', 'desc']),
    as: PropTypes__default["default"].elementType,
    bsPrefix: PropTypes__default["default"].string,
};
const TableSortLabel = React__namespace.forwardRef(({ onClick, direction = 'asc', active = false, as: Component = 'span', ...props }, ref) => {
    const getIcon = () => {
        if (!active) {
            return React__namespace.createElement("i", { className: "bi bi-arrow-down-up ms-2 align-self-center" });
        }
        return direction === 'asc' ? (React__namespace.createElement("i", { className: "bi bi-sort-up-alt ms-2 align-self-center" })) : (React__namespace.createElement("i", { className: "bi bi-sort-down ms-2 align-self-center" }));
    };
    return (React__namespace.createElement(Component, { ...props, ref: ref, role: "button", tabIndex: "0", onClick: onClick },
        props.children,
        " ",
        getIcon()));
});
TableSortLabel.displayName = 'TableSortLabel';
TableSortLabel.propTypes = propTypes$1;

const TableBody = React__namespace.forwardRef(({ as: Component = 'tbody', ...props }, ref) => {
    return React__namespace.createElement(Component, { ...props, ref: ref });
});
TableBody.displayName = 'TableBody';

const TableDataCell = React__namespace.forwardRef(({ as: Component = 'td', ...props }, ref) => {
    return (React__namespace.createElement(Component, { ...props, ref: ref }));
});
TableDataCell.displayName = 'TableDataCell';

const propTypes = {
    /**
     * @default 'table'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
   * Adds zebra-striping to any table row within the `<tbody>`.
   */
    striped: PropTypes__default["default"].bool,
    /**
     * Adds borders on all sides of the table and cells.
     */
    bordered: PropTypes__default["default"].bool,
    /**
     * Removes all borders on the table and cells, including table header.
     */
    borderless: PropTypes__default["default"].bool,
    /**
     * Enable a hover state on table rows within a `<tbody>`.
     */
    hover: PropTypes__default["default"].bool,
    /**
     * Make tables more compact by cutting cell padding in half by setting
     * size as `sm`.
     */
    size: PropTypes__default["default"].string,
    /**
     * Invert the colors of the table — with light text on dark backgrounds
     * by setting variant as `dark`.
     */
    variant: PropTypes__default["default"].string,
    /**
     * Responsive tables allow tables to be scrolled horizontally with ease.
     * Across every breakpoint, use `responsive` for horizontally
     * scrolling tables. Responsive tables are wrapped automatically in a `div`.
     * Use `responsive="sm"`, `responsive="md"`, `responsive="lg"`, or
     * `responsive="xl"` as needed to create responsive tables up to
     * a particular breakpoint. From that breakpoint and up, the table will
     * behave normally and not scroll horizontally.
     */
    responsive: PropTypes__default["default"].oneOfType([PropTypes__default["default"].bool, PropTypes__default["default"].string]),
};
const Table = React__namespace.forwardRef(({ bsPrefix, className, striped, bordered, borderless, hover, size, variant, responsive, ...props }, ref) => {
    const decoratedBsPrefix = useBootstrapPrefix(bsPrefix, 'table');
    const classes = classNames__default["default"](className, decoratedBsPrefix, variant && `${decoratedBsPrefix}-${variant}`, size && `${decoratedBsPrefix}-${size}`, striped && `${decoratedBsPrefix}-striped`, bordered && `${decoratedBsPrefix}-bordered`, borderless && `${decoratedBsPrefix}-borderless`, hover && `${decoratedBsPrefix}-hover`);
    const table = (React__namespace.createElement(SGDSWrapper, { as: "table", ...props, className: classes, ref: ref }));
    if (responsive) {
        let responsiveClass = `${decoratedBsPrefix}-responsive`;
        if (typeof responsive === 'string') {
            responsiveClass = `${responsiveClass}-${responsive}`;
        }
        return (React__namespace.createElement("div", { className: responsiveClass, tabIndex: 0 }, table));
    }
    return table;
});
Table.displayName = 'Table';
Table.propTypes = propTypes;
var Table$1 = Object.assign(Table, {
    Row: TableRow,
    Header: TableHeader,
    HeaderCell: TableHeaderCell,
    SortLabel: TableSortLabel,
    Body: TableBody,
    DataCell: TableDataCell,
});

exports.Table = Table$1;
exports.TableBody = TableBody;
exports.TableDataCell = TableDataCell;
exports.TableHeader = TableHeader;
exports.TableHeaderCell = TableHeaderCell;
exports.TableRow = TableRow;
exports.TableSortLabel = TableSortLabel;
