// @govtechsg/sgds-react/Dropdown/index.cjs.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var classNames = require('classnames');
var PropTypes = require('prop-types');
var React = require('react');
var BaseDropdown = require('@restart/ui/Dropdown');
var uncontrollable = require('uncontrollable');
var useEventCallback = require('@restart/hooks/useEventCallback');
var DropdownItem$1 = require('@restart/ui/DropdownItem');
var Anchor = require('@restart/ui/Anchor');
var DropdownMenu$1 = require('@restart/ui/DropdownMenu');
var useIsomorphicEffect = require('@restart/hooks/useIsomorphicEffect');
var useMergedRefs = require('@restart/hooks/useMergedRefs');
var warning = require('warning');
var types = require('../utils/types');
var useWrappedRefWithWarning = require('../utils/useWrappedRefWithWarning');
var DropdownToggle$1 = require('@restart/ui/DropdownToggle');
var DropdownContext$1 = require('@restart/ui/DropdownContext');
var Button$1 = require('@restart/ui/Button');
var createWithBsPrefix = require('../utils/createWithBsPrefix');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var PropTypes__default = /*#__PURE__*/_interopDefaultLegacy(PropTypes);
var React__namespace = /*#__PURE__*/_interopNamespace(React);
var BaseDropdown__default = /*#__PURE__*/_interopDefaultLegacy(BaseDropdown);
var useEventCallback__default = /*#__PURE__*/_interopDefaultLegacy(useEventCallback);
var Anchor__default = /*#__PURE__*/_interopDefaultLegacy(Anchor);
var useIsomorphicEffect__default = /*#__PURE__*/_interopDefaultLegacy(useIsomorphicEffect);
var useMergedRefs__default = /*#__PURE__*/_interopDefaultLegacy(useMergedRefs);
var warning__default = /*#__PURE__*/_interopDefaultLegacy(warning);
var useWrappedRefWithWarning__default = /*#__PURE__*/_interopDefaultLegacy(useWrappedRefWithWarning);
var DropdownContext__default = /*#__PURE__*/_interopDefaultLegacy(DropdownContext$1);
var createWithBsPrefix__default = /*#__PURE__*/_interopDefaultLegacy(createWithBsPrefix);

const DropdownContext = React__namespace.createContext({});
DropdownContext.displayName = 'DropdownContext';

const propTypes$5 = {
    /** @default 'dropdown-item' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * Highlight the menu item as active.
     */
    active: PropTypes__default["default"].bool,
    /**
     * Disable the menu item, making it unselectable.
     */
    disabled: PropTypes__default["default"].bool,
    /**
     * Value passed to the `onSelect` handler, useful for identifying the selected menu item.
     */
    eventKey: PropTypes__default["default"].oneOfType([PropTypes__default["default"].string, PropTypes__default["default"].number]),
    /**
     * HTML `href` attribute corresponding to `a.href`.
     */
    href: PropTypes__default["default"].string,
    /**
     * Callback fired when the menu item is clicked.
     */
    onClick: PropTypes__default["default"].func,
    as: PropTypes__default["default"].elementType,
};
const DropdownItem = React__namespace.forwardRef(({ bsPrefix, className, eventKey, disabled = false, onClick, active, as: Component = Anchor__default["default"], ...props }, ref) => {
    const prefix = 'dropdown-item';
    const [dropdownItemProps, meta] = DropdownItem$1.useDropdownItem({
        key: eventKey,
        href: props.href,
        disabled,
        onClick,
        active,
    });
    return (React__namespace.createElement("li", null,
        React__namespace.createElement(Component, { ...props, ...dropdownItemProps, ref: ref, className: classNames__default["default"](className, prefix, meta.isActive && 'active', disabled && 'disabled') })));
});
DropdownItem.displayName = 'DropdownItem';
DropdownItem.propTypes = propTypes$5;

const context$1 = React__namespace.createContext(null);
context$1.displayName = 'InputGroupContext';

const context = React__namespace.createContext(null);
context.displayName = 'NavbarContext';

const propTypes$4 = {
    /**
     * @default 'dropdown-menu'
     */
    bsPrefix: PropTypes__default["default"].string,
    /** Controls the visibility of the Dropdown menu  */
    show: PropTypes__default["default"].bool,
    /** Whether to render the dropdown menu in the DOM before the first time it is shown */
    renderOnMount: PropTypes__default["default"].bool,
    /** Have the dropdown switch to it's opposite placement when necessary to stay on screen. */
    flip: PropTypes__default["default"].bool,
    /**
     * Aligns the dropdown menu to the specified side of the container. You can also align
     * the menu responsively for breakpoints starting at `sm` and up. The alignment
     * direction will affect the specified breakpoint or larger.
     *
     * *Note: Using responsive alignment will disable Popper usage for positioning.*
     *
     * @type {"start"|"end"|{ sm: "start"|"end" }|{ md: "start"|"end" }|{ lg: "start"|"end" }|{ xl: "start"|"end"}|{ xxl: "start"|"end"} }
     */
    align: types.alignPropType,
    onSelect: PropTypes__default["default"].func,
    /**
     * Which event when fired outside the component will cause it to be closed
     *
     * *Note: For custom dropdown components, you will have to pass the
     * `rootCloseEvent` to `<RootCloseWrapper>` in your custom dropdown menu
     * component ([similarly to how it is implemented in `<Dropdown.Menu>`](https://github.com/react-bootstrap/react-bootstrap/blob/v0.31.5/src/DropdownMenu.js#L115-L119)).*
     */
    rootCloseEvent: PropTypes__default["default"].oneOf(['click', 'mousedown']),
    /**
     * Control the rendering of the DropdownMenu. All non-menu props
     * (listed here) are passed through to the `as` Component.
     *
     * If providing a custom, non DOM, component. the `show`, `close` and `align` props
     * are also injected and should be handled appropriately.
     */
    as: PropTypes__default["default"].elementType,
    /**
     * A set of popper options and props passed directly to Popper.
     */
    popperConfig: PropTypes__default["default"].object,
    /**
     * Menu color variant.
     *
     * Omitting this will use the default light color.
     */
    variant: PropTypes__default["default"].string,
    /** @private Provides context when used in NavDropdown component  */
    isNav: PropTypes__default["default"].bool,
};
const defaultProps$2 = {
    flip: true,
};
function getDropdownMenuPlacement(alignEnd, dropDirection, isRTL) {
    const topStart = isRTL ? 'top-end' : 'top-start';
    const topEnd = isRTL ? 'top-start' : 'top-end';
    const bottomStart = isRTL ? 'bottom-end' : 'bottom-start';
    const bottomEnd = isRTL ? 'bottom-start' : 'bottom-end';
    // const leftStart = isRTL ? 'right-start' : 'left-start';
    // const leftEnd = isRTL ? 'right-end' : 'left-end';
    const rightStart = isRTL ? 'left-start' : 'right-start';
    const rightEnd = isRTL ? 'left-end' : 'right-end';
    let placement = alignEnd ? bottomEnd : bottomStart;
    if (dropDirection === 'up')
        placement = alignEnd ? topEnd : topStart;
    else if (dropDirection === 'end')
        placement = alignEnd ? rightEnd : rightStart;
    // else if (dropDirection === 'start')
    //   placement = alignEnd ? leftEnd : leftStart;
    return placement;
}
const DropdownMenu = React__namespace.forwardRef(({ bsPrefix, className, align, rootCloseEvent, flip, show: showProps, renderOnMount, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = 'ul', popperConfig, variant, isNav, ...props }, ref) => {
    let alignEnd = false;
    const isNavbar = React.useContext(context);
    const prefix = 'dropdown-menu';
    const { align: contextAlign, drop, isRTL } = React.useContext(DropdownContext);
    align = align || contextAlign;
    const isInputGroup = React.useContext(context$1);
    const alignClasses = [];
    if (align) {
        if (typeof align === 'object') {
            const keys = Object.keys(align);
            warning__default["default"](keys.length === 1, 'There should only be 1 breakpoint when passing an object to `align`');
            if (keys.length) {
                const brkPoint = keys[0];
                const direction = align[brkPoint];
                // .dropdown-menu-end is required for responsively aligning
                // left in addition to align left classes.
                alignEnd = direction === 'start';
                alignClasses.push(`${prefix}-${brkPoint}-${direction}`);
            }
        }
        else if (align === 'end') {
            alignEnd = true;
        }
    }
    const placement = getDropdownMenuPlacement(alignEnd, drop, isRTL);
    const [menuProps, { hasShown, popper, show, toggle }] = DropdownMenu$1.useDropdownMenu({
        flip,
        rootCloseEvent,
        show: showProps,
        usePopper: !isNavbar && alignClasses.length === 0,
        offset: !isNav ? [0, 10] : undefined,
        popperConfig,
        placement,
    });
    menuProps.ref = useMergedRefs__default["default"](useWrappedRefWithWarning__default["default"](ref, 'DropdownMenu'), menuProps.ref);
    useIsomorphicEffect__default["default"](() => {
        // Popper's initial position for the menu is incorrect when
        // renderOnMount=true. Need to call update() to correct it.
        if (show)
            popper?.update();
    }, [show]);
    if (!hasShown && !renderOnMount && !isInputGroup)
        return null;
    // For custom components provide additional, non-DOM, props;
    if (typeof Component !== 'string') {
        menuProps.show = show;
        menuProps.close = () => toggle?.(false);
        menuProps.align = align;
    }
    let style = props.style;
    if (popper?.placement) {
        // we don't need the default popper style,
        // menus are display: none when not shown.
        style = { ...props.style, ...menuProps.style };
        //@ts-ignore
        props['x-placement'] = popper.placement;
    }
    return (React__namespace.createElement(Component, { ...props, ...menuProps, style: style, ...((alignClasses.length || isNavbar) && {
            'data-bs-popper': 'static',
        }), className: classNames__default["default"](className, prefix, show && 'show', alignEnd && `${prefix}-end`, variant && `${prefix}-${variant}`, ...alignClasses) }));
});
DropdownMenu.displayName = 'DropdownMenu';
DropdownMenu.propTypes = propTypes$4;
DropdownMenu.defaultProps = defaultProps$2;

const ThemeContext = React__namespace.createContext({ prefixes: {} });
({
    prefixes: PropTypes__default["default"].object,
    dir: PropTypes__default["default"].string,
});
function useBootstrapPrefix(prefix, defaultPrefix) {
    const { prefixes } = React.useContext(ThemeContext);
    return prefix || prefixes[defaultPrefix] || `${defaultPrefix}`;
}
// `sgds-${defaultPrefix}`
function useIsRTL() {
    const { dir } = React.useContext(ThemeContext);
    return dir === 'rtl';
}
const SGDSWrapper = React__namespace.forwardRef(({ as: Tag = 'div', ...props }, ref) => {
    return React__namespace.createElement(Tag, { ref: ref, ...props, className: classNames__default["default"](props.className, 'sgds') });
});

const propTypes$3 = {
    /**
     * @default 'btn'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * One or more button variant combinations
     *
     * buttons may be one of a variety of visual variants such as:
     *
     * `'primary', 'secondary', 'success', 'danger', 'warning', 'info', 'dark', 'light', 'link'`
     *
     * as well as "outline" versions (prefixed by 'outline-*')
     *
     * `'outline-primary', 'outline-secondary', 'outline-success', 'outline-danger', 'outline-warning', 'outline-info', 'outline-dark', 'outline-light'`
     */
    variant: PropTypes__default["default"].string,
    /**
     * Specifies a large or small button.
     *
     * @type ('sm'|'lg')
     */
    size: PropTypes__default["default"].string,
    /** Manually set the visual state of the button to `:active` */
    active: PropTypes__default["default"].bool,
    /**
     * Disables the Button, preventing mouse events,
     * even if the underlying component is an `<a>` element
     */
    disabled: PropTypes__default["default"].bool,
    /** Providing a `href` will render an `<a>` element, _styled_ as a button. */
    href: PropTypes__default["default"].string,
    /**
     * Defines HTML button type attribute.
     *
     * @default 'button'
     */
    type: PropTypes__default["default"].oneOf(['button', 'reset', 'submit', null]),
    as: PropTypes__default["default"].elementType,
};
const defaultProps$1 = {
    variant: 'primary',
    active: false,
    disabled: false,
};
const Button = React__namespace.forwardRef(({ as, bsPrefix, variant, size, active, className, ...props }, ref) => {
    const prefix = useBootstrapPrefix(bsPrefix, 'btn');
    const [buttonProps, { tagName }] = Button$1.useButtonProps({
        tagName: as,
        ...props,
    });
    const Component = tagName;
    return (React__namespace.createElement(SGDSWrapper, { as: Component, ...props, ...buttonProps, ref: ref, className: classNames__default["default"](className, prefix, active && 'active', variant && `${prefix}-${variant}`, size && `${prefix}-${size}`, props.href && props.disabled && 'disabled') }));
});
Button.displayName = 'Button';
Button.propTypes = propTypes$3;
Button.defaultProps = defaultProps$1;

const propTypes$2 = {
    /**
     * @default 'dropdown-toggle'
     */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * An html id attribute, necessary for assistive technologies, such as screen readers.
     * @type {string|number}
     */
    id: PropTypes__default["default"].string,
    split: PropTypes__default["default"].bool,
    as: PropTypes__default["default"].elementType,
    /**
     * to passthrough to the underlying button or whatever from DropdownButton
     * @private
     */
    childBsPrefix: PropTypes__default["default"].string,
};
const DropdownToggle = React__namespace.forwardRef(({ bsPrefix, split, className, childBsPrefix, 
// Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
as: Component = Button, ...props }, ref) => {
    const prefix = 'dropdown-toggle';
    const dropdownContext = React.useContext(DropdownContext__default["default"]);
    const isInputGroup = React.useContext(context$1);
    if (childBsPrefix !== undefined) {
        props.bsPrefix = childBsPrefix;
    }
    const [toggleProps] = DropdownToggle$1.useDropdownToggle();
    toggleProps.ref = useMergedRefs__default["default"](toggleProps.ref, useWrappedRefWithWarning__default["default"](ref, 'DropdownToggle'));
    // This intentionally forwards size and variant (if set) to the
    // underlying component, to allow it to render size and style variants.
    return (React__namespace.createElement(Component, { className: classNames__default["default"](className, prefix, split && `${prefix}-split`, !!isInputGroup && dropdownContext?.show && 'show'), variant: "outline-secondary", ...toggleProps, ...props, "aria-haspopup": "menu" }));
});
DropdownToggle.displayName = 'DropdownToggle';
DropdownToggle.propTypes = propTypes$2;

const DropdownHeader = createWithBsPrefix__default["default"]('dropdown-header', {
    defaultProps: { role: 'heading' },
});
const DropdownDivider = createWithBsPrefix__default["default"]('dropdown-divider', {
    Component: 'hr',
    defaultProps: { role: 'separator' },
});
const DropdownItemText = createWithBsPrefix__default["default"]('dropdown-item-text', {
    Component: 'span',
});

const propTypes$1 = {
    /** @default 'dropdown' */
    bsPrefix: PropTypes__default["default"].string,
    /**
     * Determines the direction and location of the Menu in relation to it's Toggle.
     *
     * @type {"end"|"up"|"down" }
     */
    drop: PropTypes__default["default"].oneOf(['up', 'end', 'down']),
    as: PropTypes__default["default"].elementType,
    align: types.alignDirection,
    /**
     * Whether or not the Dropdown is visible.
     *
     * @controllable onToggle
     */
    show: PropTypes__default["default"].bool,
    /**
     * Allow Dropdown to flip in case of an overlapping on the reference element. For more information refer to
     * Popper.js's flip [docs](https://popper.js.org/docs/v2/modifiers/flip/).
     *
     */
    flip: PropTypes__default["default"].bool,
    /**
     * A callback fired when the Dropdown wishes to change visibility. Called with the requested
     * `show` value, the DOM event, and the source that fired it: `'click'`,`'keydown'`,`'rootClose'`, or `'select'`.
     *
     * ```js
     * function(
     *   isOpen: boolean,
     *   event: SyntheticEvent,
     *   metadata: {
     *     source: 'select' | 'click' | 'rootClose' | 'keydown'
     *   }
     * ): void
     * ```
     *
     * @controllable show
     */
    onToggle: PropTypes__default["default"].func,
    /**
     * A callback fired when a menu item is selected.
     *
     * ```js
     * (eventKey: any, event: Object) => any
     * ```
     */
    onSelect: PropTypes__default["default"].func,
    /**
     * Controls the focus behavior for when the Dropdown is opened. Set to
     * `true` to always focus the first menu item, `keyboard` to focus only when
     * navigating via the keyboard, or `false` to disable completely
     *
     * The Default behavior is `false` **unless** the Menu has a `role="menu"`
     * where it will default to `keyboard` to match the recommended [ARIA Authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#menubutton).
     */
    focusFirstItemOnShow: PropTypes__default["default"].oneOf([false, true, 'keyboard']),
    /** @private */
    navbar: PropTypes__default["default"].bool,
    /**
     * Controls the auto close behaviour of the dropdown when clicking outside of
     * the button or the list.
     */
    autoClose: PropTypes__default["default"].oneOf([true, 'outside', 'inside', false]),
};
const defaultProps = {
    navbar: false,
    align: 'start',
    autoClose: true,
    focusFirstItemOnShow: 'keyboard',
};
const Dropdown = React__namespace.forwardRef((pProps, ref) => {
    const { bsPrefix, drop, show, className, align, onSelect, onToggle, focusFirstItemOnShow, 
    // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
    as: Component = 'div', navbar: _4, autoClose, ...props } = uncontrollable.useUncontrolled(pProps, { show: 'onToggle' });
    const isInputGroup = React.useContext(context$1);
    const prefix = useBootstrapPrefix(bsPrefix, 'dropdown');
    const isRTL = useIsRTL();
    const isClosingPermitted = (source) => {
        // autoClose=false only permits close on button click
        if (autoClose === false)
            return source === 'click';
        // autoClose=inside doesn't permit close on rootClose
        if (autoClose === 'inside')
            return source !== 'rootClose';
        // autoClose=outside doesn't permit close on select
        if (autoClose === 'outside')
            return source !== 'select';
        return true;
    };
    const handleToggle = useEventCallback__default["default"]((nextShow, meta) => {
        if (meta.originalEvent.currentTarget === document &&
            (meta.source !== 'keydown' ||
                meta.originalEvent.key === 'Escape'))
            meta.source = 'rootClose';
        if (isClosingPermitted(meta.source))
            onToggle?.(nextShow, meta);
    });
    const alignEnd = align === 'end';
    const placement = getDropdownMenuPlacement(alignEnd, drop, isRTL);
    const contextValue = React.useMemo(() => ({
        align,
        drop,
        isRTL,
    }), [align, drop, isRTL]);
    return (React__namespace.createElement(DropdownContext.Provider, { value: contextValue },
        React__namespace.createElement(BaseDropdown__default["default"], { placement: placement, show: show, onSelect: onSelect, onToggle: handleToggle, focusFirstItemOnShow: focusFirstItemOnShow, itemSelector: `.${prefix}-item:not(.disabled):not(:disabled)` }, isInputGroup ? (props.children) : (React__namespace.createElement(SGDSWrapper, { as: Component, ...props, ref: ref, className: classNames__default["default"](className, show && 'show', prefix, drop === 'up' && 'dropup', drop === 'end' && 'dropend') })))));
});
Dropdown.displayName = 'Dropdown';
Dropdown.propTypes = propTypes$1;
Dropdown.defaultProps = defaultProps;
var Dropdown$1 = Object.assign(Dropdown, {
    Toggle: DropdownToggle,
    Menu: DropdownMenu,
    Item: DropdownItem,
    ItemText: DropdownItemText,
    Divider: DropdownDivider,
    Header: DropdownHeader,
});

const propTypes = {
    /**
     * An html id attribute for the Toggle button, necessary for assistive technologies, such as screen readers.
     * @type {string}
     */
    id: PropTypes__default["default"].string,
    /** An `href` passed to the Toggle component */
    href: PropTypes__default["default"].string,
    /** An `onClick` handler passed to the Toggle component */
    onClick: PropTypes__default["default"].func,
    /** The content of the non-toggle Button.  */
    title: PropTypes__default["default"].node.isRequired,
    /** Disables both Buttons  */
    disabled: PropTypes__default["default"].bool,
    /**
     * Aligns the dropdown menu.
     *
     * _see [DropdownMenu](#dropdown-menu-props) for more details_
     *
     * @type {"start"|"end"|{ sm: "start"|"end" }|{ md: "start"|"end" }|{ lg: "start"|"end" }|{ xl: "start"|"end"}|{ xxl: "start"|"end"} }
     */
    align: types.alignPropType,
    /** An ARIA accessible role applied to the Menu component. When set to 'menu', The dropdown */
    menuRole: PropTypes__default["default"].string,
    /** Whether to render the dropdown menu in the DOM before the first time it is shown */
    renderMenuOnMount: PropTypes__default["default"].bool,
    /**
     *  Which event when fired outside the component will cause it to be closed.
     *
     * _see [DropdownMenu](#dropdown-menu-props) for more details_
     */
    rootCloseEvent: PropTypes__default["default"].string,
    /**
     * Menu color variant.
     *
     * Omitting this will use the default light color.
     */
    menuVariant: PropTypes__default["default"].oneOf(['dark']),
    /** @ignore */
    bsPrefix: PropTypes__default["default"].string,
    /** @ignore */
    variant: PropTypes__default["default"].string,
    /** @ignore */
    size: PropTypes__default["default"].string,
};
/**
 * A convenience component for simple or general use dropdowns. Renders a `Button` toggle and all `children`
 * are passed directly to the default `Dropdown.Menu`. This component accepts all of
 * [`Dropdown`'s props](#dropdown-props).
 *
 * _All unknown props are passed through to the `Dropdown` component._ Only
 * the Button `variant`, `size` and `bsPrefix` props are passed to the toggle,
 * along with menu-related props are passed to the `Dropdown.Menu`
 */
const DropdownButton = React__namespace.forwardRef(({ title, children, bsPrefix, rootCloseEvent, variant, size, menuRole, renderMenuOnMount, disabled, href, id, menuVariant, ...props }, ref) => (React__namespace.createElement(Dropdown$1, { ref: ref, ...props },
    React__namespace.createElement(DropdownToggle, { id: id, href: href, size: size, variant: variant, disabled: disabled, childBsPrefix: bsPrefix },
        title,
        React__namespace.createElement("i", { className: "bi bi-chevron-down" })),
    React__namespace.createElement(DropdownMenu, { role: menuRole, renderOnMount: renderMenuOnMount, rootCloseEvent: rootCloseEvent, variant: menuVariant }, children))));
DropdownButton.displayName = 'DropdownButton';
DropdownButton.propTypes = propTypes;

exports.Dropdown = Dropdown$1;
exports.DropdownButton = DropdownButton;
exports.DropdownDivider = DropdownDivider;
exports.DropdownHeader = DropdownHeader;
exports.DropdownItem = DropdownItem;
exports.DropdownItemText = DropdownItemText;
exports.DropdownMenu = DropdownMenu;
exports.DropdownToggle = DropdownToggle;
